{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f7f244d2-c114-4c6d-8b39-db2ac95e9a5a",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5145dfa-e54a-4ead-8204-3121a99187b6",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (41293311.py, line 2)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mSignatures are continuous segments of chromosomes\u001b[39m\n               ^\n\u001b[31mSyntaxError\u001b[39m\u001b[31m:\u001b[39m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# facts\n",
    "# Signatures are continuous segments of chromosomes\n",
    "# Genes can (and often do) belong to multiple signatures\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9bf62d6f-e7d1-421d-a67d-357ca5d35657",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "root = pathlib.Path(\"/data/rbg/shared/datasets/TCGA/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "64cbb3b0-795e-4bce-b8e0-a51d5c8a3875",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(133, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome  \n",
       "0              NaN       NaN   [0, 584]  \n",
       "1               No       NaN   [1, 548]  \n",
       "2              NaN       NaN  [0, 2246]  \n",
       "3              Yes       NaN  [0, 2654]  \n",
       "4              NaN       NaN   [1, 754]  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def add_outcome(elem):\n",
    "    vital_status = elem['demographic'].get('vital_status')\n",
    "\n",
    "    if vital_status == 'Dead':\n",
    "        patient_death = elem['demographic'].get('days_to_death')\n",
    "        return [1, int(patient_death)]\n",
    "\n",
    "    elif vital_status == \"Alive\":\n",
    "        follow_ups = elem['follow_ups']\n",
    "        follow_up_days = [\n",
    "            f.get('days_to_follow_up')\n",
    "            for f in follow_ups\n",
    "            if f.get('days_to_follow_up') is not None\n",
    "        ]\n",
    "        diagnosis_days = [\n",
    "            d.get('days_to_last_follow_up')\n",
    "            for d in elem['diagnoses']\n",
    "            if d.get('days_to_last_follow_up') is not None\n",
    "        ]\n",
    "        all_days = follow_up_days + diagnosis_days\n",
    "        date_of_last_checkup = max(all_days) if all_days else None\n",
    "\n",
    "        return [0, int(date_of_last_checkup)]\n",
    "    else:\n",
    "        typing.assert_never(\"vital status not in Alive/Dead\")\n",
    "\n",
    "\n",
    "def remove_sparse_features(X, threshold=0.05):\n",
    "    non_zero_fraction = (X != 0).sum(axis=0) / X.shape[0]\n",
    "    keep = non_zero_fraction > threshold\n",
    "    return X.loc[:, keep]\n",
    "\n",
    "\n",
    "def remove_perfect_correlations(X):\n",
    "    corr_matrix = X.corr().abs()\n",
    "    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "    to_drop = [column for column in upper.columns if any(upper[column] == 1)]\n",
    "    return X.drop(columns=to_drop)\n",
    "\n",
    "\n",
    "# ids = pd.read_json(\"ids/tcga_cohort.json\")[\"case_id\"]\n",
    "ids = pd.read_csv(\"sample_ids.csv\")[\"Sample\"]\n",
    "\n",
    "case_ids_raw = pd.read_json(\"./clinical-tcga-brca.json\").set_index(\"submitter_id\")\n",
    "case_ids = case_ids_raw.loc[ids]\n",
    "case_ids[\"outcome\"] = case_ids.apply(add_outcome, axis=1)\n",
    "\n",
    "case_ids = case_ids.reset_index()\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids = case_ids.drop_duplicates(subset=[\"case_id\"])\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7669c35d-d7d9-4e45-a710-0a6c4e2c6181",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1845, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Tumor Descriptor</th>\n",
       "      <th>Specimen Type</th>\n",
       "      <th>Preservation Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a</td>\n",
       "      <td>TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-C8-A27B</td>\n",
       "      <td>TCGA-C8-A27B-10A, TCGA-C8-A27B-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8ca11193-5883-4b75-9e34-f9dcbf7040a7</td>\n",
       "      <td>TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-B6-A0WW</td>\n",
       "      <td>TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d9de4dd4-6dd8-4a78-901a-35251c228ec6</td>\n",
       "      <td>TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-E9-A1R0</td>\n",
       "      <td>TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1eb79fa2-5f92-40f7-b36f-8a9844ebedf1</td>\n",
       "      <td>66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-BH-A0C1</td>\n",
       "      <td>TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42e20b03-cc2a-40e2-ba66-d34ca7890b96</td>\n",
       "      <td>TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-AC-A5EH</td>\n",
       "      <td>TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>Unknown, Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a   \n",
       "1  8ca11193-5883-4b75-9e34-f9dcbf7040a7   \n",
       "2  d9de4dd4-6dd8-4a78-901a-35251c228ec6   \n",
       "3  1eb79fa2-5f92-40f7-b36f-8a9844ebedf1   \n",
       "5  42e20b03-cc2a-40e2-ba66-d34ca7890b96   \n",
       "\n",
       "                                           File Name          Data Category  \\\n",
       "0  TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "1  TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "2  TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "3  66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...  Copy Number Variation   \n",
       "5  TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "\n",
       "                Data Type Project ID       Case ID  \\\n",
       "0     Copy Number Segment  TCGA-BRCA  TCGA-C8-A27B   \n",
       "1     Copy Number Segment  TCGA-BRCA  TCGA-B6-A0WW   \n",
       "2     Copy Number Segment  TCGA-BRCA  TCGA-E9-A1R0   \n",
       "3  Gene Level Copy Number  TCGA-BRCA  TCGA-BH-A0C1   \n",
       "5     Copy Number Segment  TCGA-BRCA  TCGA-AC-A5EH   \n",
       "\n",
       "                            Sample ID    Tissue Type         Tumor Descriptor  \\\n",
       "0  TCGA-C8-A27B-10A, TCGA-C8-A27B-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "1  TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "2  TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "3  TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "5  TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "\n",
       "                        Specimen Type Preservation Method  \n",
       "0  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "1  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "2  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "3  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "5  Solid Tissue, Peripheral Blood NOS    Unknown, Unknown  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_mapping = \"/data/rbg/shared/datasets/TCGA/Genomic/copy_number.gdc_sample_sheet.2025-07-09.tsv\"\n",
    "caseid_copynum_mapping = pd.read_csv(dna_mapping, delimiter='\\t')\n",
    "\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.split(',')\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.explode('Case ID')\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.strip()\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.drop_duplicates(subset=[\"Case ID\"]) # I'm pretty sure this is valid bc I explode only on Case ID - meaning everything else is duplicated. This whole process might actualyl be unncesary. TODO: check if all \"Case IDs pairs\" are just twins\n",
    "print(caseid_copynum_mapping.shape)\n",
    "\n",
    "# caseid_copynum_mapping.head()\n",
    "\n",
    "# stupid test to check whether any of hte records had duplicate sample IDs\n",
    "# def foo(x):\n",
    "#     a = list(map(lambda x: x.strip(), x[\"Sample ID\"].split(\",\")))\n",
    "#     print(a[0])\n",
    "#     print(a[1])\n",
    "#     # return a[0] == a[1]\n",
    "# caseid_copynum_mapping = caseid_copynum_mapping[caseid_copynum_mapping.apply(foo, axis=1)]jk:w\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ebb628da-aed9-4662-af4d-d88370d97343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(112, 20)\n"
     ]
    }
   ],
   "source": [
    "# add copynumbesr to \"IDs\" dataset\n",
    "# FIXME: figure out why 21 records are being dropped - the final result should still be 133\n",
    "print(case_ids.shape)\n",
    "data= pd.merge(\n",
    "    case_ids,\n",
    "    caseid_copynum_mapping[['Case ID', 'Sample ID', 'File ID', \"File Name\"]],  # select only the needed column\n",
    "    left_on='submitter_id',\n",
    "    right_on='Case ID',\n",
    "    how=\"inner\"\n",
    ")\n",
    "print(data.shape)\n",
    "# data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "10334a9c-f269-473c-8be3-1cfaabf8a5c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 31%|██████████████████████████████████████▊                                                                                     | 35/112 [00:13<00:30,  2.56it/s]\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Chromosome'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[32m/tmp/ipykernel_904891/524142676.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m()\u001b[39m\n\u001b[32m     28\u001b[39m \n\u001b[32m     29\u001b[39m     signatures = signatures.rename(columns={\u001b[32m0\u001b[39m: \u001b[33m\"chromosome\"\u001b[39m, \u001b[32m1\u001b[39m: \u001b[33m\"start\"\u001b[39m, \u001b[32m2\u001b[39m: \u001b[33m\"end\"\u001b[39m})\n\u001b[32m     30\u001b[39m \n\u001b[32m     31\u001b[39m \n\u001b[32m---> \u001b[39m\u001b[32m32\u001b[39m     merged = CNAs.merge(\n\u001b[32m     33\u001b[39m         signatures,\n\u001b[32m     34\u001b[39m         left_on=\u001b[33m'Chromosome'\u001b[39m,\n\u001b[32m     35\u001b[39m         right_on=\u001b[33m'chromosome'\u001b[39m,\n",
      "\u001b[32m/data/rbg/shared/projects/arpah_hs/divkov57/workspaces/conda_env/calgb_rpy/lib/python3.13/site-packages/pandas/core/frame.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[39m\n\u001b[32m  10835\u001b[39m         validate: MergeValidate | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[32m  10836\u001b[39m     ) -> DataFrame:\n\u001b[32m  10837\u001b[39m         \u001b[38;5;28;01mfrom\u001b[39;00m pandas.core.reshape.merge \u001b[38;5;28;01mimport\u001b[39;00m merge\n\u001b[32m  10838\u001b[39m \n\u001b[32m> \u001b[39m\u001b[32m10839\u001b[39m         return merge(\n\u001b[32m  10840\u001b[39m             self,\n\u001b[32m  10841\u001b[39m             right,\n\u001b[32m  10842\u001b[39m             how=how,\n",
      "\u001b[32m/data/rbg/shared/projects/arpah_hs/divkov57/workspaces/conda_env/calgb_rpy/lib/python3.13/site-packages/pandas/core/reshape/merge.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[39m\n\u001b[32m    166\u001b[39m             validate=validate,\n\u001b[32m    167\u001b[39m             copy=copy,\n\u001b[32m    168\u001b[39m         )\n\u001b[32m    169\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m170\u001b[39m         op = _MergeOperation(\n\u001b[32m    171\u001b[39m             left_df,\n\u001b[32m    172\u001b[39m             right_df,\n\u001b[32m    173\u001b[39m             how=how,\n",
      "\u001b[32m/data/rbg/shared/projects/arpah_hs/divkov57/workspaces/conda_env/calgb_rpy/lib/python3.13/site-packages/pandas/core/reshape/merge.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, indicator, validate)\u001b[39m\n\u001b[32m    790\u001b[39m             self.right_join_keys,\n\u001b[32m    791\u001b[39m             self.join_names,\n\u001b[32m    792\u001b[39m             left_drop,\n\u001b[32m    793\u001b[39m             right_drop,\n\u001b[32m--> \u001b[39m\u001b[32m794\u001b[39m         ) = self._get_merge_keys()\n\u001b[32m    795\u001b[39m \n\u001b[32m    796\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m left_drop:\n\u001b[32m    797\u001b[39m             self.left = self.left._drop_labels_or_levels(left_drop)\n",
      "\u001b[32m/data/rbg/shared/projects/arpah_hs/divkov57/workspaces/conda_env/calgb_rpy/lib/python3.13/site-packages/pandas/core/reshape/merge.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1307\u001b[39m                     \u001b[38;5;28;01mif\u001b[39;00m lk \u001b[38;5;28;01mis\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   1308\u001b[39m                         \u001b[38;5;66;03m# Then we're either Hashable or a wrong-length arraylike,\u001b[39;00m\n\u001b[32m   1309\u001b[39m                         \u001b[38;5;66;03m#  the latter of which will raise\u001b[39;00m\n\u001b[32m   1310\u001b[39m                         lk = cast(Hashable, lk)\n\u001b[32m-> \u001b[39m\u001b[32m1311\u001b[39m                         left_keys.append(left._get_label_or_level_values(lk))\n\u001b[32m   1312\u001b[39m                         join_names.append(lk)\n\u001b[32m   1313\u001b[39m                     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1314\u001b[39m                         \u001b[38;5;66;03m# work-around for merge_asof(left_index=True)\u001b[39;00m\n",
      "\u001b[32m/data/rbg/shared/projects/arpah_hs/divkov57/workspaces/conda_env/calgb_rpy/lib/python3.13/site-packages/pandas/core/generic.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, key, axis)\u001b[39m\n\u001b[32m   1907\u001b[39m             values = self.xs(key, axis=other_axes[\u001b[32m0\u001b[39m])._values\n\u001b[32m   1908\u001b[39m         \u001b[38;5;28;01melif\u001b[39;00m self._is_level_reference(key, axis=axis):\n\u001b[32m   1909\u001b[39m             values = self.axes[axis].get_level_values(key)._values\n\u001b[32m   1910\u001b[39m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1911\u001b[39m             \u001b[38;5;28;01mraise\u001b[39;00m KeyError(key)\n\u001b[32m   1912\u001b[39m \n\u001b[32m   1913\u001b[39m         \u001b[38;5;66;03m# Check for duplicates\u001b[39;00m\n\u001b[32m   1914\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m values.ndim > \u001b[32m1\u001b[39m:\n",
      "\u001b[31mKeyError\u001b[39m: 'Chromosome'"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "output = []\n",
    "for idx, elem in tqdm(list(data.iterrows())):\n",
    "    path = f\"{root}/Genomic/downloads/{elem['File ID']}/{elem['File Name']}\"\n",
    "    CNAs = pd.read_csv(path, delimiter=\"\\t\")\n",
    "\n",
    "    \n",
    "    # Example\n",
    "    # signatures = pd.DataFrame({\"signature\": [...]})\n",
    "    signatures = pd.read_excel(\"copynum_signatures.xlsx\").drop(0).rename(columns={\"Supplementary Data 2. Annotation of copy number segments\": \"signature\"})\n",
    "    \n",
    "    def extract_region(sig):\n",
    "        # Match first chr/start-end pattern\n",
    "        match = re.search(r\"(chr)?(\\d+|X|Y|M):(\\d+)-(\\d+)\", sig)\n",
    "        if match:\n",
    "            chrom = match.group(1) + match.group(2) if match.group(1) else \"chr\" + match.group(2)\n",
    "            start = int(match.group(3))\n",
    "            end = int(match.group(4))\n",
    "            return chrom, start, end\n",
    "        return None, None, None\n",
    "    \n",
    "    signatures = pd.concat([\n",
    "        signatures[\"signature\"],\n",
    "        signatures[\"signature\"].apply( lambda x: pd.Series(extract_region(x)) )\n",
    "    ], axis=1)\n",
    "    \n",
    "    signatures = signatures.rename(columns={0: \"chromosome\", 1: \"start\", 2: \"end\"})\n",
    "    \n",
    "\n",
    "    merged = CNAs.merge(\n",
    "        signatures, \n",
    "        left_on='Chromosome', \n",
    "        right_on='chromosome', \n",
    "        how='left'\n",
    "    )\n",
    "    mask = (merged['Start'] >= merged['start'])\n",
    "    \n",
    "    result = merged.loc[mask, CNAs.columns.tolist() + ['signature']]\n",
    "\n",
    "    def label_score(score):\n",
    "        if score > 0.3:\n",
    "            return \"gain\"\n",
    "        elif score < -0.3:\n",
    "            return \"loss\"\n",
    "        else:\n",
    "            return \"no-change\"\n",
    "    signature_scores = result.groupby(by=[\"signature\"])[\"Segment_Mean\"].apply(lambda x: x.mean()).apply(label_score)\n",
    "\n",
    "    signature_scores.loc[\"Case ID\"] = elem[\"submitter_id\"]\n",
    "    \n",
    "\n",
    "    output.append(signature_scores)\n",
    "        \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e4dd32c3-63b7-494f-8535-be3af8d7c479",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>signature</th>\n",
       "      <th>1.p   wholearm.chr1:1-124300000</th>\n",
       "      <th>1.q   wholearm.chr1:124300000-247249719</th>\n",
       "      <th>10.p wholearm.chr10:1-40300000</th>\n",
       "      <th>10.q wholearm.chr10:40300000-135374737</th>\n",
       "      <th>11.p wholearm.chr11:1-52900000</th>\n",
       "      <th>11.q wholearm.chr11:52900000-134452384</th>\n",
       "      <th>12.p wholearm.chr12:1-35400000</th>\n",
       "      <th>12.q wholearm.chr12:35400000-132349534</th>\n",
       "      <th>13.q wholearm.chr13:16000000-114142980</th>\n",
       "      <th>14.q wholearm.chr14:15600000-106368585</th>\n",
       "      <th>...</th>\n",
       "      <th>chr4:186684565-191273063.BeroukhimS2.4q35.2.del.</th>\n",
       "      <th>chr6:161612277-163134099.BeroukhimS2.6q26.del.PARK2</th>\n",
       "      <th>chr7:144118814-148066271.BeroukhimS2.7q35.del.CNTNAP2</th>\n",
       "      <th>chr7:156893473-158821424.BeroukhimS2.7q36.3.del.PTPRN2</th>\n",
       "      <th>chr9:137859478-140273252.BeroukhimS2.9q34.3.amp.</th>\n",
       "      <th>chrX:148267454-148753845.BeroukhimS5..amp.</th>\n",
       "      <th>chrX:152729030-154913754.BeroukhimS2.Xq28.amp.</th>\n",
       "      <th>chr20:56249879-56298730.BeroukhimS5..amp.</th>\n",
       "      <th>chr20:59868877-61137250.BeroukhimS5..amp.</th>\n",
       "      <th>chr20:61329497-62435964.BeroukhimS2.2q13.33.amp.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>...</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>...</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>...</td>\n",
       "      <td>gain</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>...</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>loss</td>\n",
       "      <td>loss</td>\n",
       "      <td>no-change</td>\n",
       "      <td>no-change</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "      <td>gain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 532 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "signature 1.p   wholearm.chr1:1-124300000  \\\n",
       "0                               no-change   \n",
       "1                               no-change   \n",
       "2                               no-change   \n",
       "3                                    gain   \n",
       "4                               no-change   \n",
       "\n",
       "signature 1.q   wholearm.chr1:124300000-247249719  \\\n",
       "0                                       no-change   \n",
       "1                                       no-change   \n",
       "2                                       no-change   \n",
       "3                                            gain   \n",
       "4                                       no-change   \n",
       "\n",
       "signature 10.p wholearm.chr10:1-40300000  \\\n",
       "0                                   loss   \n",
       "1                              no-change   \n",
       "2                                   loss   \n",
       "3                                   gain   \n",
       "4                                   gain   \n",
       "\n",
       "signature 10.q wholearm.chr10:40300000-135374737  \\\n",
       "0                                           loss   \n",
       "1                                           loss   \n",
       "2                                           loss   \n",
       "3                                           gain   \n",
       "4                                      no-change   \n",
       "\n",
       "signature 11.p wholearm.chr11:1-52900000  \\\n",
       "0                              no-change   \n",
       "1                                   loss   \n",
       "2                              no-change   \n",
       "3                              no-change   \n",
       "4                              no-change   \n",
       "\n",
       "signature 11.q wholearm.chr11:52900000-134452384  \\\n",
       "0                                      no-change   \n",
       "1                                           loss   \n",
       "2                                      no-change   \n",
       "3                                      no-change   \n",
       "4                                           loss   \n",
       "\n",
       "signature 12.p wholearm.chr12:1-35400000  \\\n",
       "0                                   gain   \n",
       "1                              no-change   \n",
       "2                              no-change   \n",
       "3                                   gain   \n",
       "4                              no-change   \n",
       "\n",
       "signature 12.q wholearm.chr12:35400000-132349534  \\\n",
       "0                                      no-change   \n",
       "1                                           loss   \n",
       "2                                      no-change   \n",
       "3                                           gain   \n",
       "4                                      no-change   \n",
       "\n",
       "signature 13.q wholearm.chr13:16000000-114142980  \\\n",
       "0                                           loss   \n",
       "1                                           loss   \n",
       "2                                           loss   \n",
       "3                                      no-change   \n",
       "4                                           loss   \n",
       "\n",
       "signature 14.q wholearm.chr14:15600000-106368585  ...  \\\n",
       "0                                           loss  ...   \n",
       "1                                      no-change  ...   \n",
       "2                                           loss  ...   \n",
       "3                                           loss  ...   \n",
       "4                                      no-change  ...   \n",
       "\n",
       "signature chr4:186684565-191273063.BeroukhimS2.4q35.2.del.  \\\n",
       "0                                                      NaN   \n",
       "1                                                     loss   \n",
       "2                                                     loss   \n",
       "3                                                     gain   \n",
       "4                                                no-change   \n",
       "\n",
       "signature chr6:161612277-163134099.BeroukhimS2.6q26.del.PARK2  \\\n",
       "0                                                        NaN    \n",
       "1                                                  no-change    \n",
       "2                                                  no-change    \n",
       "3                                                       loss    \n",
       "4                                                  no-change    \n",
       "\n",
       "signature chr7:144118814-148066271.BeroukhimS2.7q35.del.CNTNAP2  \\\n",
       "0                                                        NaN      \n",
       "1                                                  no-change      \n",
       "2                                                  no-change      \n",
       "3                                                  no-change      \n",
       "4                                                  no-change      \n",
       "\n",
       "signature chr7:156893473-158821424.BeroukhimS2.7q36.3.del.PTPRN2  \\\n",
       "0                                                        NaN       \n",
       "1                                                       loss       \n",
       "2                                                       gain       \n",
       "3                                                        NaN       \n",
       "4                                                       loss       \n",
       "\n",
       "signature chr9:137859478-140273252.BeroukhimS2.9q34.3.amp.  \\\n",
       "0                                                      NaN   \n",
       "1                                                     loss   \n",
       "2                                                no-change   \n",
       "3                                                      NaN   \n",
       "4                                                     loss   \n",
       "\n",
       "signature chrX:148267454-148753845.BeroukhimS5..amp.  \\\n",
       "0                                                NaN   \n",
       "1                                          no-change   \n",
       "2                                                NaN   \n",
       "3                                          no-change   \n",
       "4                                          no-change   \n",
       "\n",
       "signature chrX:152729030-154913754.BeroukhimS2.Xq28.amp.  \\\n",
       "0                                                    NaN   \n",
       "1                                              no-change   \n",
       "2                                                    NaN   \n",
       "3                                              no-change   \n",
       "4                                              no-change   \n",
       "\n",
       "signature chr20:56249879-56298730.BeroukhimS5..amp.  \\\n",
       "0                                               NaN   \n",
       "1                                               NaN   \n",
       "2                                         no-change   \n",
       "3                                              gain   \n",
       "4                                              gain   \n",
       "\n",
       "signature chr20:59868877-61137250.BeroukhimS5..amp.  \\\n",
       "0                                               NaN   \n",
       "1                                               NaN   \n",
       "2                                         no-change   \n",
       "3                                              gain   \n",
       "4                                              gain   \n",
       "\n",
       "signature chr20:61329497-62435964.BeroukhimS2.2q13.33.amp.  \n",
       "0                                                      NaN  \n",
       "1                                                      NaN  \n",
       "2                                                no-change  \n",
       "3                                                no-change  \n",
       "4                                                     gain  \n",
       "\n",
       "[5 rows x 532 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cna_data = pd.DataFrame(output).reset_index(drop=True)\n",
    "\n",
    "cna_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "16a14d90-a5fc-4cd1-baa1-6bd9a913c579",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     TCGA-A1-A0SP\n",
       "1     TCGA-A2-A04P\n",
       "2     TCGA-A2-A04T\n",
       "3     TCGA-A2-A04U\n",
       "4     TCGA-A2-A0CM\n",
       "5     TCGA-A2-A0D0\n",
       "6     TCGA-A2-A0D2\n",
       "7     TCGA-A2-A0ST\n",
       "8     TCGA-A2-A0T0\n",
       "9     TCGA-A2-A0YE\n",
       "10    TCGA-A2-A0YM\n",
       "11    TCGA-A2-A3XT\n",
       "12    TCGA-A2-A3XU\n",
       "13    TCGA-A2-A3XX\n",
       "14    TCGA-A2-A3XY\n",
       "15    TCGA-A7-A0DA\n",
       "16    TCGA-A7-A26I\n",
       "17    TCGA-A7-A4SD\n",
       "18    TCGA-A7-A4SE\n",
       "19    TCGA-A7-A6VV\n",
       "20    TCGA-A7-A6VW\n",
       "21    TCGA-A7-A6VY\n",
       "22    TCGA-A8-A07O\n",
       "23    TCGA-A8-A07R\n",
       "24    TCGA-A8-A08R\n",
       "25    TCGA-AC-A8OQ\n",
       "26    TCGA-AN-A04D\n",
       "27    TCGA-AN-A0AR\n",
       "28    TCGA-AN-A0AT\n",
       "29    TCGA-AN-A0FX\n",
       "30    TCGA-AN-A0G0\n",
       "31    TCGA-AN-A0XU\n",
       "32    TCGA-AO-A0J6\n",
       "33    TCGA-AO-A124\n",
       "34    TCGA-AO-A128\n",
       "Name: Case ID, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cna_data[\"Case ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "39490943-d9ed-4a60-addf-05487a035a55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 2)\n",
      "(35, 532)\n",
      "(35, 534)\n"
     ]
    }
   ],
   "source": [
    "case_ids = case_ids[[\"submitter_id\", \"outcome\"]]\n",
    "\n",
    "df = case_ids.merge(\n",
    "    cna_data,\n",
    "    right_on=[\"Case ID\"],\n",
    "    left_on=[\"submitter_id\"],\n",
    "    how=\"inner\"\n",
    ")\n",
    "\n",
    "print(case_ids.shape)\n",
    "print(cna_data.shape)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2986855d-dbdb-4b60-886d-c7efc672f613",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outcome</th>\n",
       "      <th>1.p   wholearm.chr1:1-124300000</th>\n",
       "      <th>1.q   wholearm.chr1:124300000-247249719</th>\n",
       "      <th>10.p wholearm.chr10:1-40300000</th>\n",
       "      <th>10.q wholearm.chr10:40300000-135374737</th>\n",
       "      <th>11.p wholearm.chr11:1-52900000</th>\n",
       "      <th>11.q wholearm.chr11:52900000-134452384</th>\n",
       "      <th>12.p wholearm.chr12:1-35400000</th>\n",
       "      <th>12.q wholearm.chr12:35400000-132349534</th>\n",
       "      <th>13.q wholearm.chr13:16000000-114142980</th>\n",
       "      <th>...</th>\n",
       "      <th>chr4:186684565-191273063.BeroukhimS2.4q35.2.del.</th>\n",
       "      <th>chr6:161612277-163134099.BeroukhimS2.6q26.del.PARK2</th>\n",
       "      <th>chr7:144118814-148066271.BeroukhimS2.7q35.del.CNTNAP2</th>\n",
       "      <th>chr7:156893473-158821424.BeroukhimS2.7q36.3.del.PTPRN2</th>\n",
       "      <th>chr9:137859478-140273252.BeroukhimS2.9q34.3.amp.</th>\n",
       "      <th>chrX:148267454-148753845.BeroukhimS5..amp.</th>\n",
       "      <th>chrX:152729030-154913754.BeroukhimS2.Xq28.amp.</th>\n",
       "      <th>chr20:56249879-56298730.BeroukhimS5..amp.</th>\n",
       "      <th>chr20:59868877-61137250.BeroukhimS5..amp.</th>\n",
       "      <th>chr20:61329497-62435964.BeroukhimS2.2q13.33.amp.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[0, 584]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[1, 548]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[0, 2246]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[0, 2654]</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[1, 754]</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 532 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     outcome  1.p   wholearm.chr1:1-124300000  \\\n",
       "0   [0, 584]                                2   \n",
       "1   [1, 548]                                2   \n",
       "2  [0, 2246]                                2   \n",
       "3  [0, 2654]                                0   \n",
       "4   [1, 754]                                2   \n",
       "\n",
       "   1.q   wholearm.chr1:124300000-247249719  10.p wholearm.chr10:1-40300000  \\\n",
       "0                                        2                               1   \n",
       "1                                        2                               2   \n",
       "2                                        2                               1   \n",
       "3                                        0                               0   \n",
       "4                                        2                               0   \n",
       "\n",
       "   10.q wholearm.chr10:40300000-135374737  11.p wholearm.chr11:1-52900000  \\\n",
       "0                                       1                               2   \n",
       "1                                       1                               1   \n",
       "2                                       1                               2   \n",
       "3                                       0                               2   \n",
       "4                                       2                               2   \n",
       "\n",
       "   11.q wholearm.chr11:52900000-134452384  12.p wholearm.chr12:1-35400000  \\\n",
       "0                                       2                               0   \n",
       "1                                       1                               2   \n",
       "2                                       2                               2   \n",
       "3                                       2                               0   \n",
       "4                                       1                               2   \n",
       "\n",
       "   12.q wholearm.chr12:35400000-132349534  \\\n",
       "0                                       2   \n",
       "1                                       1   \n",
       "2                                       2   \n",
       "3                                       0   \n",
       "4                                       2   \n",
       "\n",
       "   13.q wholearm.chr13:16000000-114142980  ...  \\\n",
       "0                                       1  ...   \n",
       "1                                       1  ...   \n",
       "2                                       1  ...   \n",
       "3                                       2  ...   \n",
       "4                                       1  ...   \n",
       "\n",
       "   chr4:186684565-191273063.BeroukhimS2.4q35.2.del.  \\\n",
       "0                                                 3   \n",
       "1                                                 1   \n",
       "2                                                 1   \n",
       "3                                                 0   \n",
       "4                                                 2   \n",
       "\n",
       "   chr6:161612277-163134099.BeroukhimS2.6q26.del.PARK2  \\\n",
       "0                                                  3     \n",
       "1                                                  2     \n",
       "2                                                  2     \n",
       "3                                                  1     \n",
       "4                                                  2     \n",
       "\n",
       "   chr7:144118814-148066271.BeroukhimS2.7q35.del.CNTNAP2  \\\n",
       "0                                                  3       \n",
       "1                                                  2       \n",
       "2                                                  2       \n",
       "3                                                  2       \n",
       "4                                                  2       \n",
       "\n",
       "   chr7:156893473-158821424.BeroukhimS2.7q36.3.del.PTPRN2  \\\n",
       "0                                                  3        \n",
       "1                                                  1        \n",
       "2                                                  0        \n",
       "3                                                  3        \n",
       "4                                                  1        \n",
       "\n",
       "   chr9:137859478-140273252.BeroukhimS2.9q34.3.amp.  \\\n",
       "0                                                 3   \n",
       "1                                                 1   \n",
       "2                                                 2   \n",
       "3                                                 3   \n",
       "4                                                 1   \n",
       "\n",
       "   chrX:148267454-148753845.BeroukhimS5..amp.  \\\n",
       "0                                           3   \n",
       "1                                           2   \n",
       "2                                           3   \n",
       "3                                           2   \n",
       "4                                           2   \n",
       "\n",
       "   chrX:152729030-154913754.BeroukhimS2.Xq28.amp.  \\\n",
       "0                                               3   \n",
       "1                                               2   \n",
       "2                                               3   \n",
       "3                                               2   \n",
       "4                                               2   \n",
       "\n",
       "   chr20:56249879-56298730.BeroukhimS5..amp.  \\\n",
       "0                                          3   \n",
       "1                                          3   \n",
       "2                                          2   \n",
       "3                                          0   \n",
       "4                                          0   \n",
       "\n",
       "   chr20:59868877-61137250.BeroukhimS5..amp.  \\\n",
       "0                                          3   \n",
       "1                                          3   \n",
       "2                                          2   \n",
       "3                                          0   \n",
       "4                                          0   \n",
       "\n",
       "   chr20:61329497-62435964.BeroukhimS2.2q13.33.amp.  \n",
       "0                                                 3  \n",
       "1                                                 3  \n",
       "2                                                 2  \n",
       "3                                                 2  \n",
       "4                                                 0  \n",
       "\n",
       "[5 rows x 532 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = df.drop([\"Case ID\", \"submitter_id\"], axis=1)\n",
    "# X = df.drop([\"outcome\"], axis=1)\n",
    "# y = df[\"outcome\"]\n",
    "\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "le = LabelEncoder()\n",
    "for col in df:\n",
    "    if col != \"outcome\":\n",
    "        df[col] = le.fit_transform(df[col])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b80051d5-9f97-423e-bf6f-95154a2ac542",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import multiprocessing\n",
    "multiprocessing.cpu_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "efe6e710-ceda-40aa-b38b-fceeea02fa33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "  0%|                                                                                                                                     | 0/180 [00:00<?, ?it/s]\u001b[A\u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "!!! skipped -- no uncencored records\n",
      "Skipping fold - no comparable pairs in test set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=20)]: Using backend LokyBackend with 20 concurrent workers.\n",
      "\n",
      "\n",
      "  1%|▋                                                                                                                            | 1/180 [00:00<01:52,  1.60it/s]\u001b[A\u001b[A\n",
      "\n",
      " 22%|███████████████████████████▌                                                                                                | 40/180 [00:06<00:21,  6.55it/s]\u001b[A\u001b[A[Parallel(n_jobs=20)]: Done  10 tasks      | elapsed:    6.7s\n",
      "\n",
      "\n",
      " 33%|█████████████████████████████████████████▎                                                                                  | 60/180 [00:09<00:18,  6.64it/s]\u001b[A\u001b[A\n",
      "\n",
      " 44%|███████████████████████████████████████████████████████                                                                     | 80/180 [00:12<00:14,  6.75it/s]\u001b[A\u001b[A\n",
      "\n",
      " 56%|████████████████████████████████████████████████████████████████████▎                                                      | 100/180 [00:15<00:11,  6.69it/s]\u001b[A\u001b[A\n",
      "\n",
      " 67%|██████████████████████████████████████████████████████████████████████████████████                                         | 120/180 [00:18<00:09,  6.65it/s]\u001b[A\u001b[A\n",
      "\n",
      " 78%|███████████████████████████████████████████████████████████████████████████████████████████████▋                           | 140/180 [00:21<00:06,  6.65it/s]\u001b[A\u001b[A\n",
      "\n",
      " 89%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████▎             | 160/180 [00:24<00:03,  6.65it/s]\u001b[A\u001b[A\n",
      "\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 180/180 [00:27<00:00,  6.59it/s]\u001b[A\u001b[A\n",
      "[Parallel(n_jobs=20)]: Done 180 out of 180 | elapsed:   33.2s finished\n"
     ]
    }
   ],
   "source": [
    "# resampling test\n",
    "from sklearn.model_selection import ShuffleSplit, ParameterGrid\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sksurv.linear_model import CoxnetSurvivalAnalysis\n",
    "from joblib import Parallel, delayed\n",
    "import tqdm\n",
    "import warnings\n",
    "\n",
    "from sksurv.util import Surv\n",
    "from sklearn.utils.validation import validate_data\n",
    "\n",
    "l1_ratios = np.arange(0.1, 1.0, 0.1)\n",
    "lambdas = np.logspace(2, -2, 20)\n",
    "\n",
    "param_grid = list(ParameterGrid({\n",
    "    'l1_ratio': l1_ratios,\n",
    "    'lambda': lambdas\n",
    "}))\n",
    "print(len(param_grid))\n",
    "\n",
    "\n",
    "cv = ShuffleSplit(n_splits=25, test_size=0.25, random_state=42)\n",
    "gtmp = 0\n",
    "\n",
    "from sksurv.metrics import NoComparablePairException\n",
    "\n",
    "def train_parameters(cv, param):\n",
    "    l1_ratio = param[\"l1_ratio\"]\n",
    "    lambda_ = param[\"lambda\"]\n",
    "\n",
    "    score_sum = 0\n",
    "    score_count = 0\n",
    "\n",
    "    for train_idx, test_idx in cv.split(df):\n",
    "        train_df = df.iloc[train_idx]\n",
    "        test_df = df.iloc[test_idx]\n",
    "\n",
    "        X_train = train_df.drop(\"outcome\", axis=1).apply(pd.to_numeric, errors='raise').to_numpy()\n",
    "        X_test = test_df.drop(\"outcome\", axis=1).apply(pd.to_numeric, errors='raise').to_numpy()\n",
    "\n",
    "        train_outcome_df = pd.DataFrame(train_df[\"outcome\"].tolist(), columns=[\"event\", \"duration\"]).astype({\"event\": bool, \"duration\": float})\n",
    "        y_train = Surv.from_dataframe(\"event\", \"duration\", train_outcome_df)\n",
    "\n",
    "        test_outcome_df = pd.DataFrame(test_df[\"outcome\"].tolist(), columns=[\"event\", \"duration\"]).astype({\"event\": bool, \"duration\": float})\n",
    "        y_test = Surv.from_dataframe(\"event\", \"duration\", test_outcome_df)\n",
    "\n",
    "        cox = CoxnetSurvivalAnalysis(\n",
    "            l1_ratio=l1_ratio,\n",
    "            alphas=np.atleast_1d(lambda_).astype(float),\n",
    "            max_iter=1000\n",
    "        )\n",
    "        with warnings.catch_warnings():\n",
    "            warnings.simplefilter(\"ignore\", UserWarning)\n",
    "            cox.fit(X=X_train, y=y_train)\n",
    "\n",
    "        try:\n",
    "            if pd.DataFrame(y_test)[\"event\"].any():\n",
    "                score = cox.score(X_test, y_test)\n",
    "                score_sum += score\n",
    "                score_count += 1\n",
    "            else:\n",
    "                print(\"!!! skipped -- no uncencored records\")\n",
    "        except NoComparablePairException:\n",
    "            print(\"Skipping fold - no comparable pairs in test set\")\n",
    "\n",
    "    return {\n",
    "        'l1_ratio': l1_ratio,\n",
    "        'lambda': lambda_,\n",
    "        'score': score_sum / score_count if score_count > 0 else float('nan')\n",
    "    }\n",
    "\n",
    "\n",
    "gtmp = Parallel(backend=\"loky\", n_jobs=-1, verbose=1)(delayed(train_parameters)(cv, param) for param in tqdm.tqdm(param_grid))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ba36e5d1-6baf-4f1d-836a-7adf3a6dce5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n",
      "[{'l1_ratio': np.float64(0.1), 'lambda': np.float64(100.0), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(61.58482110660264), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(37.926901907322495), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(23.357214690901223), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(14.38449888287663), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(8.858667904100827), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(5.45559478116852), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(3.3598182862837827), 'score': np.float64(0.5)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(2.06913808111479), 'score': np.float64(0.41228070175438597)}, {'l1_ratio': np.float64(0.1), 'lambda': np.float64(1.2742749857031341), 'score': np.float64(0.45560871876661346)}]\n"
     ]
    }
   ],
   "source": [
    "print(len(gtmp))\n",
    "print(gtmp[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "808a9165-8e26-4db7-bfce-43cf9e9cb006",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_params = max(gtmp, key=lambda x: x['score'])\n",
    "l1_ratio, _lambda, score = best_params['l1_ratio'], best_params['lambda'], best_params['score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4b64639a-6a8a-41d8-ae33-b9590a0600db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_904891/1548947657.py:12: ConvergenceWarning: Optimization terminated early, you might want to increase the number of iterations (max_iter=1000).\n",
      "  model.fit(X=X, y=y)\n"
     ]
    }
   ],
   "source": [
    "X = df.drop(\"outcome\", axis=1).apply(pd.to_numeric, errors='raise').to_numpy()\n",
    "\n",
    "_y= pd.DataFrame(df[\"outcome\"].tolist(), columns=[\"event\", \"duration\"]).astype({\"event\": bool, \"duration\": float})\n",
    "y = Surv.from_dataframe(\"event\", \"duration\", _y)\n",
    "\n",
    "model = CoxnetSurvivalAnalysis(\n",
    "            l1_ratio=l1_ratio,\n",
    "            alphas=np.atleast_1d(_lambda).astype(float),\n",
    "            max_iter=1000\n",
    "        )\n",
    "\n",
    "model.fit(X=X, y=y)\n",
    "\n",
    "coef_series = pd.Series(model.coef_.reshape(-1), index=df.drop(\"outcome\", axis=1).columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "0e0feb2c-0785-48f2-87d1-50f4c8695c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr14:93528266-103946058.BeroukhimS5..amp.               0.000486\n",
       "chr14:92773649-95250286.Basal.14q32.13-99.del            0.000973\n",
       "chr10:89467202-90419015.BeroukhimS2.1q23.31.del.PTEN     0.001311\n",
       "6.q   wholearm.chr6:60500000-170899992                   0.001596\n",
       "chr7:144118814-148066271.BeroukhimS2.7q35.del.CNTNAP2    0.002943\n",
       "                                                           ...   \n",
       "chr20:41621022-45719023.LumB.20q13.12-166.amp           -0.174658\n",
       "chr9:137859478-140273252.BeroukhimS2.9q34.3.amp.        -0.226767\n",
       "chr9:130999928-132303780.BeroukhimS5..del.              -0.244604\n",
       "chr8:140458177-146274826.BeroukhimS2.8q24.3.amp.        -0.331828\n",
       "chr20:61329497-62435964.BeroukhimS2.2q13.33.amp.        -0.339865\n",
       "Length: 118, dtype: float64"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coef_series[coef_series != 0].sort_values(key=lambda x: abs(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ede128-59d6-4405-ba64-da0375eafa0e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
