{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ea8168f4-92f8-4e4d-9b93-41d6e2bcd740",
   "metadata": {},
   "source": [
    "# facts\n",
    "Signatures are continuous segments of chromosomes\n",
    "Genes can (and often do) belong to multiple signatures\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9efecdf-b857-4974-b68b-978bc67074c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "root = pathlib.Path(\"/data/rbg/shared/datasets/TCGA/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13e383d4-3504-45ec-a358-23d8dc253caf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(133, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome  \n",
       "0              NaN       NaN   [0, 584]  \n",
       "1               No       NaN   [1, 548]  \n",
       "2              NaN       NaN  [0, 2246]  \n",
       "3              Yes       NaN  [0, 2654]  \n",
       "4              NaN       NaN   [1, 754]  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def add_outcome(elem):\n",
    "    vital_status = elem['demographic'].get('vital_status')\n",
    "\n",
    "    if vital_status == 'Dead':\n",
    "        patient_death = elem['demographic'].get('days_to_death')\n",
    "        return [1, int(patient_death)]\n",
    "\n",
    "    elif vital_status == \"Alive\":\n",
    "        follow_ups = elem['follow_ups']\n",
    "        follow_up_days = [\n",
    "            f.get('days_to_follow_up')\n",
    "            for f in follow_ups\n",
    "            if f.get('days_to_follow_up') is not None\n",
    "        ]\n",
    "        diagnosis_days = [\n",
    "            d.get('days_to_last_follow_up')\n",
    "            for d in elem['diagnoses']\n",
    "            if d.get('days_to_last_follow_up') is not None\n",
    "        ]\n",
    "        all_days = follow_up_days + diagnosis_days\n",
    "        date_of_last_checkup = max(all_days) if all_days else None\n",
    "\n",
    "        return [0, int(date_of_last_checkup)]\n",
    "    else:\n",
    "        typing.assert_never(\"vital status not in Alive/Dead\")\n",
    "\n",
    "\n",
    "def remove_sparse_features(X, threshold=0.05):\n",
    "    non_zero_fraction = (X != 0).sum(axis=0) / X.shape[0]\n",
    "    keep = non_zero_fraction > threshold\n",
    "    return X.loc[:, keep]\n",
    "\n",
    "\n",
    "def remove_perfect_correlations(X):\n",
    "    corr_matrix = X.corr().abs()\n",
    "    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "    to_drop = [column for column in upper.columns if any(upper[column] == 1)]\n",
    "    return X.drop(columns=to_drop)\n",
    "\n",
    "\n",
    "# ids = pd.read_json(\"ids/tcga_cohort.json\")[\"case_id\"]\n",
    "ids = pd.read_csv(\"sample_ids.csv\")[\"Sample\"]\n",
    "\n",
    "case_ids_raw = pd.read_json(\"./clinical-tcga-brca.json\").set_index(\"submitter_id\")\n",
    "case_ids = case_ids_raw.loc[ids]\n",
    "case_ids[\"outcome\"] = case_ids.apply(add_outcome, axis=1)\n",
    "\n",
    "case_ids = case_ids.reset_index()\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids = case_ids.drop_duplicates(subset=[\"case_id\"])\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f255396-1170-4260-ab44-34755d0bae5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2690, 11)\n",
      "(5380, 11)\n",
      "(1845, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Tumor Descriptor</th>\n",
       "      <th>Specimen Type</th>\n",
       "      <th>Preservation Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a</td>\n",
       "      <td>TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-C8-A27B</td>\n",
       "      <td>TCGA-C8-A27B-10A, TCGA-C8-A27B-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8ca11193-5883-4b75-9e34-f9dcbf7040a7</td>\n",
       "      <td>TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-B6-A0WW</td>\n",
       "      <td>TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d9de4dd4-6dd8-4a78-901a-35251c228ec6</td>\n",
       "      <td>TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-E9-A1R0</td>\n",
       "      <td>TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1eb79fa2-5f92-40f7-b36f-8a9844ebedf1</td>\n",
       "      <td>66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-BH-A0C1</td>\n",
       "      <td>TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42e20b03-cc2a-40e2-ba66-d34ca7890b96</td>\n",
       "      <td>TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-AC-A5EH</td>\n",
       "      <td>TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>Unknown, Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a   \n",
       "1  8ca11193-5883-4b75-9e34-f9dcbf7040a7   \n",
       "2  d9de4dd4-6dd8-4a78-901a-35251c228ec6   \n",
       "3  1eb79fa2-5f92-40f7-b36f-8a9844ebedf1   \n",
       "5  42e20b03-cc2a-40e2-ba66-d34ca7890b96   \n",
       "\n",
       "                                           File Name          Data Category  \\\n",
       "0  TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "1  TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "2  TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "3  66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...  Copy Number Variation   \n",
       "5  TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "\n",
       "                Data Type Project ID       Case ID  \\\n",
       "0     Copy Number Segment  TCGA-BRCA  TCGA-C8-A27B   \n",
       "1     Copy Number Segment  TCGA-BRCA  TCGA-B6-A0WW   \n",
       "2     Copy Number Segment  TCGA-BRCA  TCGA-E9-A1R0   \n",
       "3  Gene Level Copy Number  TCGA-BRCA  TCGA-BH-A0C1   \n",
       "5     Copy Number Segment  TCGA-BRCA  TCGA-AC-A5EH   \n",
       "\n",
       "                            Sample ID    Tissue Type         Tumor Descriptor  \\\n",
       "0  TCGA-C8-A27B-10A, TCGA-C8-A27B-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "1  TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "2  TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "3  TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "5  TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "\n",
       "                        Specimen Type Preservation Method  \n",
       "0  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "1  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "2  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "3  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "5  Solid Tissue, Peripheral Blood NOS    Unknown, Unknown  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_mapping = \"/data/rbg/shared/datasets/TCGA/Genomic/copy_number.gdc_sample_sheet.2025-07-09.tsv\"\n",
    "caseid_copynum_mapping = pd.read_csv(dna_mapping, delimiter='\\t')\n",
    "\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.split(',')\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.explode('Case ID')\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.strip()\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.drop_duplicates(subset=[\"Case ID\"]) # I'm pretty sure this is valid bc I explode only on Case ID - meaning everything else is duplicated. This whole process might actualyl be unncesary. TODO: check if all \"Case IDs pairs\" are just twins\n",
    "print(caseid_copynum_mapping.shape)\n",
    "\n",
    "caseid_copynum_mapping.head()\n",
    "\n",
    "# stupid test to check whether any of hte records had duplicate sample IDs\n",
    "# def foo(x):\n",
    "#     a = list(map(lambda x: x.strip(), x[\"Sample ID\"].split(\",\")))\n",
    "#     print(a[0])\n",
    "#     print(a[1])\n",
    "#     # return a[0] == a[1]\n",
    "# caseid_copynum_mapping = caseid_copynum_mapping[caseid_copynum_mapping.apply(foo, axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c5ae89c6-a920-4472-af53-a156e7b4e45f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(112, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>TCGA-A1-A0SP-01A, TCGA-A1-A0SP-10A</td>\n",
       "      <td>4dfa8f5b-e9de-46d5-9aa5-15a2dc535ec8</td>\n",
       "      <td>TCGA-A1-A0SP-01A-11D-A893-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>TCGA-A2-A04P-10A, TCGA-A2-A04P-01A</td>\n",
       "      <td>e175ae5e-11da-4bd1-bfb4-5919c85a72fc</td>\n",
       "      <td>TCGA-A2-A04P-01A-31D-A89H-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>TCGA-A2-A04T-01A, TCGA-A2-A04T-10A</td>\n",
       "      <td>017a228b-af6a-46a0-840f-f50d64f4e434</td>\n",
       "      <td>b752b444-f033-4be4-9d24-e5e80b4181af_wgs_gdc_r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>TCGA-A2-A04U-10A, TCGA-A2-A04U-01A</td>\n",
       "      <td>bf33ef54-ff2a-49f6-a24e-6d6052df4e73</td>\n",
       "      <td>TCGA-A2-A04U-01A-11D-A89H-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>TCGA-A2-A0CM-01A, TCGA-A2-A0CM-10A</td>\n",
       "      <td>a59ae3ed-af09-4cb2-bf60-518f85798d53</td>\n",
       "      <td>TCGA-A2-A0CM-01A-31D-A890-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome       Case ID  \\\n",
       "0              NaN       NaN   [0, 584]  TCGA-A1-A0SP   \n",
       "1               No       NaN   [1, 548]  TCGA-A2-A04P   \n",
       "2              NaN       NaN  [0, 2246]  TCGA-A2-A04T   \n",
       "3              Yes       NaN  [0, 2654]  TCGA-A2-A04U   \n",
       "4              NaN       NaN   [1, 754]  TCGA-A2-A0CM   \n",
       "\n",
       "                            Sample ID                               File ID  \\\n",
       "0  TCGA-A1-A0SP-01A, TCGA-A1-A0SP-10A  4dfa8f5b-e9de-46d5-9aa5-15a2dc535ec8   \n",
       "1  TCGA-A2-A04P-10A, TCGA-A2-A04P-01A  e175ae5e-11da-4bd1-bfb4-5919c85a72fc   \n",
       "2  TCGA-A2-A04T-01A, TCGA-A2-A04T-10A  017a228b-af6a-46a0-840f-f50d64f4e434   \n",
       "3  TCGA-A2-A04U-10A, TCGA-A2-A04U-01A  bf33ef54-ff2a-49f6-a24e-6d6052df4e73   \n",
       "4  TCGA-A2-A0CM-01A, TCGA-A2-A0CM-10A  a59ae3ed-af09-4cb2-bf60-518f85798d53   \n",
       "\n",
       "                                           File Name  \n",
       "0  TCGA-A1-A0SP-01A-11D-A893-36.WholeGenome.RP-16...  \n",
       "1  TCGA-A2-A04P-01A-31D-A89H-36.WholeGenome.RP-16...  \n",
       "2  b752b444-f033-4be4-9d24-e5e80b4181af_wgs_gdc_r...  \n",
       "3  TCGA-A2-A04U-01A-11D-A89H-36.WholeGenome.RP-16...  \n",
       "4  TCGA-A2-A0CM-01A-31D-A890-36.WholeGenome.RP-16...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add copynumbesr to \"IDs\" dataset\n",
    "# FIXME: figure out why 21 records are being dropped - the final result should still be 133\n",
    "print(case_ids.shape)\n",
    "data= pd.merge(\n",
    "    case_ids,\n",
    "    caseid_copynum_mapping[['Case ID', 'Sample ID', 'File ID', \"File Name\"]],  # select only the needed column\n",
    "    left_on='submitter_id',\n",
    "    right_on='Case ID',\n",
    "    how=\"inner\"\n",
    ")\n",
    "print(data.shape)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fd3ff3c0-13cf-45ed-8a49-3bf869824169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3723001</td>\n",
       "      <td>5795000</td>\n",
       "      <td>2061</td>\n",
       "      <td>-0.080328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5795001</td>\n",
       "      <td>7432000</td>\n",
       "      <td>1634</td>\n",
       "      <td>-0.028879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7432001</td>\n",
       "      <td>9532000</td>\n",
       "      <td>2094</td>\n",
       "      <td>-0.017117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9532001</td>\n",
       "      <td>9540000</td>\n",
       "      <td>8</td>\n",
       "      <td>-0.181266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3737</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26403001</td>\n",
       "      <td>26442000</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.650024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26638001</td>\n",
       "      <td>56708000</td>\n",
       "      <td>6</td>\n",
       "      <td>-9.341383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3739</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56821001</td>\n",
       "      <td>56882000</td>\n",
       "      <td>61</td>\n",
       "      <td>-0.278835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3740</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882001</td>\n",
       "      <td>56887000</td>\n",
       "      <td>5</td>\n",
       "      <td>-2.367349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3741</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrM</td>\n",
       "      <td>1</td>\n",
       "      <td>16569</td>\n",
       "      <td>17</td>\n",
       "      <td>-0.301116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3742 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GDC_Aliquot_ID Chromosome     Start       End  Num_Probes  \\\n",
       "0     TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000        2754   \n",
       "1     TCGA-A2-A0YM-01A-11D-A894-36       chr1   3723001   5795000        2061   \n",
       "2     TCGA-A2-A0YM-01A-11D-A894-36       chr1   5795001   7432000        1634   \n",
       "3     TCGA-A2-A0YM-01A-11D-A894-36       chr1   7432001   9532000        2094   \n",
       "4     TCGA-A2-A0YM-01A-11D-A894-36       chr1   9532001   9540000           8   \n",
       "...                            ...        ...       ...       ...         ...   \n",
       "3737  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26403001  26442000           5   \n",
       "3738  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26638001  56708000           6   \n",
       "3739  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56821001  56882000          61   \n",
       "3740  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56882001  56887000           5   \n",
       "3741  TCGA-A2-A0YM-01A-11D-A894-36       chrM         1     16569          17   \n",
       "\n",
       "      Segment_Mean  \n",
       "0        -0.048360  \n",
       "1        -0.080328  \n",
       "2        -0.028879  \n",
       "3        -0.017117  \n",
       "4        -0.181266  \n",
       "...            ...  \n",
       "3737     -1.650024  \n",
       "3738     -9.341383  \n",
       "3739     -0.278835  \n",
       "3740     -2.367349  \n",
       "3741     -0.301116  \n",
       "\n",
       "[3742 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elem = data.iloc[10]\n",
    "path = f\"{root}/Genomic/downloads/{elem['File ID']}/{elem['File Name']}\"\n",
    "CNAs = pd.read_csv(path, delimiter=\"\\t\")\n",
    "CNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "439f708d-3dc7-4a9c-adce-54a40f177962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2:1-15244284.BeroukhimS2.2p25.3.del.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>1</td>\n",
       "      <td>15244284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.p   wholearm.chr2:1-93300000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>1</td>\n",
       "      <td>93300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15:56490060-57176541.Basal.15q22.1-107.del</td>\n",
       "      <td>chr15</td>\n",
       "      <td>56490060</td>\n",
       "      <td>57176541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22:23139372-23249329.BeroukhimS5..amp.</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23139372</td>\n",
       "      <td>23249329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr12:97551177-99047626.BeroukhimS2.12q23.1.de...</td>\n",
       "      <td>chr12</td>\n",
       "      <td>97551177</td>\n",
       "      <td>99047626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           signature chromosome     start  \\\n",
       "1            chr2:1-15244284.BeroukhimS2.2p25.3.del.       chr2         1   \n",
       "2                     2.p   wholearm.chr2:1-93300000       chr2         1   \n",
       "3      chr15:56490060-57176541.Basal.15q22.1-107.del      chr15  56490060   \n",
       "4          chr22:23139372-23249329.BeroukhimS5..amp.      chr22  23139372   \n",
       "5  chr12:97551177-99047626.BeroukhimS2.12q23.1.de...      chr12  97551177   \n",
       "\n",
       "        end  \n",
       "1  15244284  \n",
       "2  93300000  \n",
       "3  57176541  \n",
       "4  23249329  \n",
       "5  99047626  "
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Example\n",
    "# signatures = pd.DataFrame({\"signature\": [...]})\n",
    "signatures = pd.read_excel(\"copynum_signatures.xlsx\").drop(0).rename(columns={\"Supplementary Data 2. Annotation of copy number segments\": \"signature\"})\n",
    "\n",
    "def extract_region(sig):\n",
    "    # Match first chr/start-end pattern\n",
    "    match = re.search(r\"(chr)?(\\d+|X|Y|M):(\\d+)-(\\d+)\", sig)\n",
    "    if match:\n",
    "        chrom = match.group(1) + match.group(2) if match.group(1) else \"chr\" + match.group(2)\n",
    "        start = int(match.group(3))\n",
    "        end = int(match.group(4))\n",
    "        return chrom, start, end\n",
    "    return None, None, None\n",
    "\n",
    "signatures = pd.concat([\n",
    "    signatures[\"signature\"],\n",
    "    signatures[\"signature\"].apply( lambda x: pd.Series(extract_region(x)) )\n",
    "], axis=1)\n",
    "\n",
    "signatures = signatures.rename(columns={0: \"chromosome\", 1: \"start\", 2: \"end\"})\n",
    "\n",
    "signatures.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "057f486b-22ed-4621-affb-bfd1e4a50c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = CNAs.df.merge(\n",
    "    signatures, \n",
    "    left_on='Chromosome', \n",
    "    right_on='chromosome', \n",
    "    how='left'\n",
    ")\n",
    "mask = (merged['Start'] >= merged['start'])\n",
    "\n",
    "result = merged.loc[mask, CNAs.columns.tolist() + ['signature']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "ce50b672-cfa5-4a35-aa7f-7a4bba2bd133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "      <th>signature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>1.p   wholearm.chr1:1-124300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3723001</td>\n",
       "      <td>5795000</td>\n",
       "      <td>2061</td>\n",
       "      <td>-0.080328</td>\n",
       "      <td>1.p   wholearm.chr1:1-124300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5795001</td>\n",
       "      <td>7432000</td>\n",
       "      <td>1634</td>\n",
       "      <td>-0.028879</td>\n",
       "      <td>chr1:3756302-6867390.BeroukhimS2.1p36.31.del.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5795001</td>\n",
       "      <td>7432000</td>\n",
       "      <td>1634</td>\n",
       "      <td>-0.028879</td>\n",
       "      <td>1.p   wholearm.chr1:1-124300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7432001</td>\n",
       "      <td>9532000</td>\n",
       "      <td>2094</td>\n",
       "      <td>-0.017117</td>\n",
       "      <td>chr1:3756302-6867390.BeroukhimS2.1p36.31.del.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82592</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155967001</td>\n",
       "      <td>156026000</td>\n",
       "      <td>43</td>\n",
       "      <td>-0.374156</td>\n",
       "      <td>X.qwholearm.chrX:60000000-154913754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82593</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155967001</td>\n",
       "      <td>156026000</td>\n",
       "      <td>43</td>\n",
       "      <td>-0.374156</td>\n",
       "      <td>chrX:31041721-34564697.BeroukhimS2.Xp21.2.del.DMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82594</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155967001</td>\n",
       "      <td>156026000</td>\n",
       "      <td>43</td>\n",
       "      <td>-0.374156</td>\n",
       "      <td>chrX:8707411-9925292.BeroukhimS5..del.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82595</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155967001</td>\n",
       "      <td>156026000</td>\n",
       "      <td>43</td>\n",
       "      <td>-0.374156</td>\n",
       "      <td>chrX:148267454-148753845.BeroukhimS5..amp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82596</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155967001</td>\n",
       "      <td>156026000</td>\n",
       "      <td>43</td>\n",
       "      <td>-0.374156</td>\n",
       "      <td>chrX:66436234-67090514.BeroukhimS2.Xq12.amp.AR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43268 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GDC_Aliquot_ID Chromosome      Start        End  \\\n",
       "25     TCGA-A2-A0YM-01A-11D-A894-36       chr1      17001    3723000   \n",
       "53     TCGA-A2-A0YM-01A-11D-A894-36       chr1    3723001    5795000   \n",
       "57     TCGA-A2-A0YM-01A-11D-A894-36       chr1    5795001    7432000   \n",
       "81     TCGA-A2-A0YM-01A-11D-A894-36       chr1    5795001    7432000   \n",
       "85     TCGA-A2-A0YM-01A-11D-A894-36       chr1    7432001    9532000   \n",
       "...                             ...        ...        ...        ...   \n",
       "82592  TCGA-A2-A0YM-01A-11D-A894-36       chrX  155967001  156026000   \n",
       "82593  TCGA-A2-A0YM-01A-11D-A894-36       chrX  155967001  156026000   \n",
       "82594  TCGA-A2-A0YM-01A-11D-A894-36       chrX  155967001  156026000   \n",
       "82595  TCGA-A2-A0YM-01A-11D-A894-36       chrX  155967001  156026000   \n",
       "82596  TCGA-A2-A0YM-01A-11D-A894-36       chrX  155967001  156026000   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \\\n",
       "25           2754     -0.048360   \n",
       "53           2061     -0.080328   \n",
       "57           1634     -0.028879   \n",
       "81           1634     -0.028879   \n",
       "85           2094     -0.017117   \n",
       "...           ...           ...   \n",
       "82592          43     -0.374156   \n",
       "82593          43     -0.374156   \n",
       "82594          43     -0.374156   \n",
       "82595          43     -0.374156   \n",
       "82596          43     -0.374156   \n",
       "\n",
       "                                               signature  \n",
       "25                       1.p   wholearm.chr1:1-124300000  \n",
       "53                       1.p   wholearm.chr1:1-124300000  \n",
       "57         chr1:3756302-6867390.BeroukhimS2.1p36.31.del.  \n",
       "81                       1.p   wholearm.chr1:1-124300000  \n",
       "85         chr1:3756302-6867390.BeroukhimS2.1p36.31.del.  \n",
       "...                                                  ...  \n",
       "82592                X.qwholearm.chrX:60000000-154913754  \n",
       "82593  chrX:31041721-34564697.BeroukhimS2.Xp21.2.del.DMD  \n",
       "82594             chrX:8707411-9925292.BeroukhimS5..del.  \n",
       "82595         chrX:148267454-148753845.BeroukhimS5..amp.  \n",
       "82596     chrX:66436234-67090514.BeroukhimS2.Xq12.amp.AR  \n",
       "\n",
       "[43268 rows x 7 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "a39a3ad6-df14-43cc-98df-5ac0b0f11590",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "signature\n",
       "1.p   wholearm.chr1:1-124300000                   [-0.04836, -0.080328, -0.028879, -0.017117, -0...\n",
       "1.q   wholearm.chr1:124300000-247249719           [1.125795, -3.244125, 1.782699, 0.448627, 1.30...\n",
       "10.p wholearm.chr10:1-40300000                    [0.282643, 0.220832, 0.302888, 0.250967, 0.263...\n",
       "10.q wholearm.chr10:40300000-135374737            [0.470294, -0.080945, -0.786112, -0.08005, 0.5...\n",
       "11.p wholearm.chr11:1-52900000                    [-0.073111, -0.3931, -0.108993, -0.088036, -0....\n",
       "                                                                        ...                        \n",
       "chrX:43505580-46847125.BeroukhimS5..amp.          [0.582329, -0.374064, 0.554671, -0.376845, -0....\n",
       "chrX:66436234-67090514.BeroukhimS2.Xq12.amp.AR    [0.582329, -0.374064, 0.554671, -0.376845, -0....\n",
       "chrX:6695466-8345492.BeroukhimS5..del.            [0.031391, 0.582329, -0.374064, 0.554671, -0.3...\n",
       "chrX:69545961-73353657.BeroukhimS5..amp.          [0.582329, -0.374064, 0.554671, -0.376845, -0....\n",
       "chrX:8707411-9925292.BeroukhimS5..del.            [0.031391, 0.582329, -0.374064, 0.554671, -0.3...\n",
       "Name: Segment_Mean, Length: 528, dtype: object"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.groupby(by=[\"signature\"])[\"Segment_Mean\"].apply(lambda x: x.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "cfb69bd8-39b6-4492-832f-937da999cde0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "signature\n",
       "1.p   wholearm.chr1:1-124300000                   0.562351\n",
       "1.q   wholearm.chr1:124300000-247249719           0.512755\n",
       "10.p wholearm.chr10:1-40300000                    0.043622\n",
       "10.q wholearm.chr10:40300000-135374737           -0.006903\n",
       "11.p wholearm.chr11:1-52900000                   -0.221676\n",
       "                                                    ...   \n",
       "chrX:43505580-46847125.BeroukhimS5..amp.         -0.184523\n",
       "chrX:66436234-67090514.BeroukhimS2.Xq12.amp.AR   -0.184523\n",
       "chrX:6695466-8345492.BeroukhimS5..del.           -0.164894\n",
       "chrX:69545961-73353657.BeroukhimS5..amp.         -0.184523\n",
       "chrX:8707411-9925292.BeroukhimS5..del.           -0.164894\n",
       "Name: Segment_Mean, Length: 528, dtype: float64"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signature_scores = result.groupby(by=[\"signature\"])[\"Segment_Mean\"].apply(lambda x: x.mean())\n",
    "signature_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "1bcc2eda-8e0d-4993-9e27-009c5b5d7a6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.22348484848484848)"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(signature_scores > 0.3).sum() / signature_scores.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "77e864c8-5b18-4189-86c9-82de68146eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "signature_scores.loc[\"Case ID\"] = elem[\"submitter_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "ff96e098-9e52-4b8a-bb92-2fba180cb400",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "signature\n",
       "1.p   wholearm.chr1:1-124300000                       0.562351\n",
       "1.q   wholearm.chr1:124300000-247249719               0.512755\n",
       "10.p wholearm.chr10:1-40300000                        0.043622\n",
       "10.q wholearm.chr10:40300000-135374737               -0.006903\n",
       "11.p wholearm.chr11:1-52900000                       -0.221676\n",
       "                                                      ...     \n",
       "chrX:66436234-67090514.BeroukhimS2.Xq12.amp.AR       -0.184523\n",
       "chrX:6695466-8345492.BeroukhimS5..del.               -0.164894\n",
       "chrX:69545961-73353657.BeroukhimS5..amp.             -0.184523\n",
       "chrX:8707411-9925292.BeroukhimS5..del.               -0.164894\n",
       "Case ID                                           TCGA-A2-A0YM\n",
       "Name: Segment_Mean, Length: 529, dtype: object"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signature_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba81be52-8448-41d4-8909-8e049d3faaf7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
