{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ea8168f4-92f8-4e4d-9b93-41d6e2bcd740",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9efecdf-b857-4974-b68b-978bc67074c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "root = pathlib.Path(\"/data/rbg/shared/datasets/TCGA/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13e383d4-3504-45ec-a358-23d8dc253caf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(133, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome  \n",
       "0              NaN       NaN   [0, 584]  \n",
       "1               No       NaN   [1, 548]  \n",
       "2              NaN       NaN  [0, 2246]  \n",
       "3              Yes       NaN  [0, 2654]  \n",
       "4              NaN       NaN   [1, 754]  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def add_outcome(elem):\n",
    "    vital_status = elem['demographic'].get('vital_status')\n",
    "\n",
    "    if vital_status == 'Dead':\n",
    "        patient_death = elem['demographic'].get('days_to_death')\n",
    "        return [1, int(patient_death)]\n",
    "\n",
    "    elif vital_status == \"Alive\":\n",
    "        follow_ups = elem['follow_ups']\n",
    "        follow_up_days = [\n",
    "            f.get('days_to_follow_up')\n",
    "            for f in follow_ups\n",
    "            if f.get('days_to_follow_up') is not None\n",
    "        ]\n",
    "        diagnosis_days = [\n",
    "            d.get('days_to_last_follow_up')\n",
    "            for d in elem['diagnoses']\n",
    "            if d.get('days_to_last_follow_up') is not None\n",
    "        ]\n",
    "        all_days = follow_up_days + diagnosis_days\n",
    "        date_of_last_checkup = max(all_days) if all_days else None\n",
    "\n",
    "        return [0, int(date_of_last_checkup)]\n",
    "    else:\n",
    "        typing.assert_never(\"vital status not in Alive/Dead\")\n",
    "\n",
    "\n",
    "def remove_sparse_features(X, threshold=0.05):\n",
    "    non_zero_fraction = (X != 0).sum(axis=0) / X.shape[0]\n",
    "    keep = non_zero_fraction > threshold\n",
    "    return X.loc[:, keep]\n",
    "\n",
    "\n",
    "def remove_perfect_correlations(X):\n",
    "    corr_matrix = X.corr().abs()\n",
    "    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "    to_drop = [column for column in upper.columns if any(upper[column] == 1)]\n",
    "    return X.drop(columns=to_drop)\n",
    "\n",
    "\n",
    "# ids = pd.read_json(\"ids/tcga_cohort.json\")[\"case_id\"]\n",
    "ids = pd.read_csv(\"sample_ids.csv\")[\"Sample\"]\n",
    "\n",
    "case_ids_raw = pd.read_json(\"./clinical-tcga-brca.json\").set_index(\"submitter_id\")\n",
    "case_ids = case_ids_raw.loc[ids]\n",
    "case_ids[\"outcome\"] = case_ids.apply(add_outcome, axis=1)\n",
    "\n",
    "case_ids = case_ids.reset_index()\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids = case_ids.drop_duplicates(subset=[\"case_id\"])\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f255396-1170-4260-ab44-34755d0bae5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2690, 11)\n",
      "(5380, 11)\n",
      "(1845, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Tumor Descriptor</th>\n",
       "      <th>Specimen Type</th>\n",
       "      <th>Preservation Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a</td>\n",
       "      <td>TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-C8-A27B</td>\n",
       "      <td>TCGA-C8-A27B-10A, TCGA-C8-A27B-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8ca11193-5883-4b75-9e34-f9dcbf7040a7</td>\n",
       "      <td>TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-B6-A0WW</td>\n",
       "      <td>TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d9de4dd4-6dd8-4a78-901a-35251c228ec6</td>\n",
       "      <td>TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-E9-A1R0</td>\n",
       "      <td>TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1eb79fa2-5f92-40f7-b36f-8a9844ebedf1</td>\n",
       "      <td>66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-BH-A0C1</td>\n",
       "      <td>TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42e20b03-cc2a-40e2-ba66-d34ca7890b96</td>\n",
       "      <td>TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Copy Number Segment</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-AC-A5EH</td>\n",
       "      <td>TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>Unknown, Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  b260bc0a-bf82-4fe8-b1ca-1ba43a4d741a   \n",
       "1  8ca11193-5883-4b75-9e34-f9dcbf7040a7   \n",
       "2  d9de4dd4-6dd8-4a78-901a-35251c228ec6   \n",
       "3  1eb79fa2-5f92-40f7-b36f-8a9844ebedf1   \n",
       "5  42e20b03-cc2a-40e2-ba66-d34ca7890b96   \n",
       "\n",
       "                                           File Name          Data Category  \\\n",
       "0  TCGA-C8-A27B-01A-11D-A89A-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "1  TCGA-B6-A0WW-01A-11D-A893-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "2  TCGA-E9-A1R0-01A-22D-A898-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "3  66a9e6d2-6dea-4475-ab39-7b5ffa3d9faf.wgs.ASCAT...  Copy Number Variation   \n",
       "5  TCGA-AC-A5EH-01A-11D-A89D-36.WholeGenome.RP-16...  Copy Number Variation   \n",
       "\n",
       "                Data Type Project ID       Case ID  \\\n",
       "0     Copy Number Segment  TCGA-BRCA  TCGA-C8-A27B   \n",
       "1     Copy Number Segment  TCGA-BRCA  TCGA-B6-A0WW   \n",
       "2     Copy Number Segment  TCGA-BRCA  TCGA-E9-A1R0   \n",
       "3  Gene Level Copy Number  TCGA-BRCA  TCGA-BH-A0C1   \n",
       "5     Copy Number Segment  TCGA-BRCA  TCGA-AC-A5EH   \n",
       "\n",
       "                            Sample ID    Tissue Type         Tumor Descriptor  \\\n",
       "0  TCGA-C8-A27B-10A, TCGA-C8-A27B-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "1  TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "2  TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A  Normal, Tumor  Not Applicable, Primary   \n",
       "3  TCGA-BH-A0C1-01B, TCGA-BH-A0C1-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "5  TCGA-AC-A5EH-01A, TCGA-AC-A5EH-10A  Tumor, Normal  Primary, Not Applicable   \n",
       "\n",
       "                        Specimen Type Preservation Method  \n",
       "0  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "1  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "2  Peripheral Blood NOS, Solid Tissue        Unknown, OCT  \n",
       "3  Solid Tissue, Peripheral Blood NOS        OCT, Unknown  \n",
       "5  Solid Tissue, Peripheral Blood NOS    Unknown, Unknown  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_mapping = \"/data/rbg/shared/datasets/TCGA/Genomic/copy_number.gdc_sample_sheet.2025-07-09.tsv\"\n",
    "caseid_copynum_mapping = pd.read_csv(dna_mapping, delimiter='\\t')\n",
    "\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.split(',')\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.explode('Case ID')\n",
    "caseid_copynum_mapping['Case ID'] = caseid_copynum_mapping['Case ID'].str.strip()\n",
    "# print(caseid_copynum_mapping.shape)\n",
    "caseid_copynum_mapping = caseid_copynum_mapping.drop_duplicates(subset=[\"Case ID\"]) # I'm pretty sure this is valid bc I explode only on Case ID - meaning everything else is duplicated. This whole process might actualyl be unncesary. TODO: check if all \"Case IDs pairs\" are just twins\n",
    "print(caseid_copynum_mapping.shape)\n",
    "\n",
    "caseid_copynum_mapping.head()\n",
    "\n",
    "# stupid test to check whether any of hte records had duplicate sample IDs\n",
    "# def foo(x):\n",
    "#     a = list(map(lambda x: x.strip(), x[\"Sample ID\"].split(\",\")))\n",
    "#     print(a[0])\n",
    "#     print(a[1])\n",
    "#     # return a[0] == a[1]\n",
    "# caseid_copynum_mapping = caseid_copynum_mapping[caseid_copynum_mapping.apply(foo, axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c5ae89c6-a920-4472-af53-a156e7b4e45f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(112, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>TCGA-A1-A0SP-01A, TCGA-A1-A0SP-10A</td>\n",
       "      <td>4dfa8f5b-e9de-46d5-9aa5-15a2dc535ec8</td>\n",
       "      <td>TCGA-A1-A0SP-01A-11D-A893-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>TCGA-A2-A04P-10A, TCGA-A2-A04P-01A</td>\n",
       "      <td>e175ae5e-11da-4bd1-bfb4-5919c85a72fc</td>\n",
       "      <td>TCGA-A2-A04P-01A-31D-A89H-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>TCGA-A2-A04T-01A, TCGA-A2-A04T-10A</td>\n",
       "      <td>017a228b-af6a-46a0-840f-f50d64f4e434</td>\n",
       "      <td>b752b444-f033-4be4-9d24-e5e80b4181af_wgs_gdc_r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>TCGA-A2-A04U-10A, TCGA-A2-A04U-01A</td>\n",
       "      <td>bf33ef54-ff2a-49f6-a24e-6d6052df4e73</td>\n",
       "      <td>TCGA-A2-A04U-01A-11D-A89H-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>TCGA-A2-A0CM-01A, TCGA-A2-A0CM-10A</td>\n",
       "      <td>a59ae3ed-af09-4cb2-bf60-518f85798d53</td>\n",
       "      <td>TCGA-A2-A0CM-01A-31D-A890-36.WholeGenome.RP-16...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome       Case ID  \\\n",
       "0              NaN       NaN   [0, 584]  TCGA-A1-A0SP   \n",
       "1               No       NaN   [1, 548]  TCGA-A2-A04P   \n",
       "2              NaN       NaN  [0, 2246]  TCGA-A2-A04T   \n",
       "3              Yes       NaN  [0, 2654]  TCGA-A2-A04U   \n",
       "4              NaN       NaN   [1, 754]  TCGA-A2-A0CM   \n",
       "\n",
       "                            Sample ID                               File ID  \\\n",
       "0  TCGA-A1-A0SP-01A, TCGA-A1-A0SP-10A  4dfa8f5b-e9de-46d5-9aa5-15a2dc535ec8   \n",
       "1  TCGA-A2-A04P-10A, TCGA-A2-A04P-01A  e175ae5e-11da-4bd1-bfb4-5919c85a72fc   \n",
       "2  TCGA-A2-A04T-01A, TCGA-A2-A04T-10A  017a228b-af6a-46a0-840f-f50d64f4e434   \n",
       "3  TCGA-A2-A04U-10A, TCGA-A2-A04U-01A  bf33ef54-ff2a-49f6-a24e-6d6052df4e73   \n",
       "4  TCGA-A2-A0CM-01A, TCGA-A2-A0CM-10A  a59ae3ed-af09-4cb2-bf60-518f85798d53   \n",
       "\n",
       "                                           File Name  \n",
       "0  TCGA-A1-A0SP-01A-11D-A893-36.WholeGenome.RP-16...  \n",
       "1  TCGA-A2-A04P-01A-31D-A89H-36.WholeGenome.RP-16...  \n",
       "2  b752b444-f033-4be4-9d24-e5e80b4181af_wgs_gdc_r...  \n",
       "3  TCGA-A2-A04U-01A-11D-A89H-36.WholeGenome.RP-16...  \n",
       "4  TCGA-A2-A0CM-01A-31D-A890-36.WholeGenome.RP-16...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add copynumbesr to \"IDs\" dataset\n",
    "# FIXME: figure out why 21 records are being dropped - the final result should still be 133\n",
    "print(case_ids.shape)\n",
    "data= pd.merge(\n",
    "    case_ids,\n",
    "    caseid_copynum_mapping[['Case ID', 'Sample ID', 'File ID', \"File Name\"]],  # select only the needed column\n",
    "    left_on='submitter_id',\n",
    "    right_on='Case ID',\n",
    "    how=\"inner\"\n",
    ")\n",
    "print(data.shape)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fd3ff3c0-13cf-45ed-8a49-3bf869824169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3723001</td>\n",
       "      <td>5795000</td>\n",
       "      <td>2061</td>\n",
       "      <td>-0.080328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5795001</td>\n",
       "      <td>7432000</td>\n",
       "      <td>1634</td>\n",
       "      <td>-0.028879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7432001</td>\n",
       "      <td>9532000</td>\n",
       "      <td>2094</td>\n",
       "      <td>-0.017117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9532001</td>\n",
       "      <td>9540000</td>\n",
       "      <td>8</td>\n",
       "      <td>-0.181266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3737</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26403001</td>\n",
       "      <td>26442000</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.650024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26638001</td>\n",
       "      <td>56708000</td>\n",
       "      <td>6</td>\n",
       "      <td>-9.341383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3739</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56821001</td>\n",
       "      <td>56882000</td>\n",
       "      <td>61</td>\n",
       "      <td>-0.278835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3740</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882001</td>\n",
       "      <td>56887000</td>\n",
       "      <td>5</td>\n",
       "      <td>-2.367349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3741</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrM</td>\n",
       "      <td>1</td>\n",
       "      <td>16569</td>\n",
       "      <td>17</td>\n",
       "      <td>-0.301116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3742 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GDC_Aliquot_ID Chromosome     Start       End  Num_Probes  \\\n",
       "0     TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000        2754   \n",
       "1     TCGA-A2-A0YM-01A-11D-A894-36       chr1   3723001   5795000        2061   \n",
       "2     TCGA-A2-A0YM-01A-11D-A894-36       chr1   5795001   7432000        1634   \n",
       "3     TCGA-A2-A0YM-01A-11D-A894-36       chr1   7432001   9532000        2094   \n",
       "4     TCGA-A2-A0YM-01A-11D-A894-36       chr1   9532001   9540000           8   \n",
       "...                            ...        ...       ...       ...         ...   \n",
       "3737  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26403001  26442000           5   \n",
       "3738  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26638001  56708000           6   \n",
       "3739  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56821001  56882000          61   \n",
       "3740  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56882001  56887000           5   \n",
       "3741  TCGA-A2-A0YM-01A-11D-A894-36       chrM         1     16569          17   \n",
       "\n",
       "      Segment_Mean  \n",
       "0        -0.048360  \n",
       "1        -0.080328  \n",
       "2        -0.028879  \n",
       "3        -0.017117  \n",
       "4        -0.181266  \n",
       "...            ...  \n",
       "3737     -1.650024  \n",
       "3738     -9.341383  \n",
       "3739     -0.278835  \n",
       "3740     -2.367349  \n",
       "3741     -0.301116  \n",
       "\n",
       "[3742 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elem = data.iloc[10]\n",
    "path = f\"{root}/Genomic/downloads/{elem['File ID']}/{elem['File Name']}\"\n",
    "CNAs = pd.read_csv(path, delimiter=\"\\t\")\n",
    "CNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "87c63a9c-a4a7-490a-9b53-14e3fffc8e1d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(62307, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62302</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>14941746</td>\n",
       "      <td>14942715</td>\n",
       "      <td>AC006989.2</td>\n",
       "      <td>ENSG00000224518.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62303</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>15348662</td>\n",
       "      <td>15456074</td>\n",
       "      <td>PUDPP1</td>\n",
       "      <td>ENSG00000234620.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62304</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>15547216</td>\n",
       "      <td>15593331</td>\n",
       "      <td>STSP1</td>\n",
       "      <td>ENSG00000227166.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62305</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26403001</td>\n",
       "      <td>26442000</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.650024</td>\n",
       "      <td>26409815</td>\n",
       "      <td>26420535</td>\n",
       "      <td>ANKRD36P1</td>\n",
       "      <td>ENSG00000188399.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62306</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56821001</td>\n",
       "      <td>56882000</td>\n",
       "      <td>61</td>\n",
       "      <td>-0.278835</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000235857.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62307 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GDC_Aliquot_ID Chromosome     Start       End  \\\n",
       "0      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "1      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "2      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "3      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "4      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "...                             ...        ...       ...       ...   \n",
       "62302  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62303  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62304  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62305  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26403001  26442000   \n",
       "62306  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56821001  56882000   \n",
       "\n",
       "       Num_Probes  Segment_Mean   Start_b     End_b    gene_name  \\\n",
       "0            2754     -0.048360     14404     29570       WASH7P   \n",
       "1            2754     -0.048360     17369     17436    MIR6859-1   \n",
       "2            2754     -0.048360     29554     31109  MIR1302-2HG   \n",
       "3            2754     -0.048360     30366     30503    MIR1302-2   \n",
       "4            2754     -0.048360     34554     36081      FAM138A   \n",
       "...           ...           ...       ...       ...          ...   \n",
       "62302          11     -0.629715  14941746  14942715   AC006989.2   \n",
       "62303          11     -0.629715  15348662  15456074       PUDPP1   \n",
       "62304          11     -0.629715  15547216  15593331        STSP1   \n",
       "62305           5     -1.650024  26409815  26420535    ANKRD36P1   \n",
       "62306          61     -0.278835  56855244  56855488      CTBP2P1   \n",
       "\n",
       "                 gene_id  \n",
       "0      ENSG00000227232.5  \n",
       "1      ENSG00000278267.1  \n",
       "2      ENSG00000243485.5  \n",
       "3      ENSG00000284332.1  \n",
       "4      ENSG00000237613.2  \n",
       "...                  ...  \n",
       "62302  ENSG00000224518.2  \n",
       "62303  ENSG00000234620.1  \n",
       "62304  ENSG00000227166.1  \n",
       "62305  ENSG00000188399.5  \n",
       "62306  ENSG00000235857.1  \n",
       "\n",
       "[62307 rows x 10 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Joins CNAs (CHR, start, end) w/ a generated (CHR, start, end -> gene_name) mapping based on overlaping ranges\n",
    "# several genes can exist in a range -> increase df size\n",
    "\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "\n",
    "if isinstance(CNAs, pd.DataFrame):\n",
    "    CNAs = pr.PyRanges(CNAs)\n",
    "elif isinstance(CNAs, pr.PyRanges):\n",
    "    CNAs = CNAs\n",
    "else:\n",
    "    typing.assert_never(\"impossible CNAs datatype\")\n",
    "\n",
    "def parse_gtf(filepath):\n",
    "    # GTF files are tab-separated, and comments start with '#'\n",
    "    gtf = pd.read_csv(filepath, sep='\\t', comment='#', header=None,\n",
    "                      names=[\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"])\n",
    "    \n",
    "    genes = gtf[gtf[\"feature\"] == \"gene\"].copy()\n",
    "    \n",
    "    def extract_attr(attr_str, key):\n",
    "        import re\n",
    "        match = re.search(f'{key} \"([^\"]+)\"', attr_str)\n",
    "        return match.group(1) if match else None\n",
    "    \n",
    "    genes[\"gene_name\"] = genes[\"attribute\"].apply(lambda x: extract_attr(x, \"gene_name\"))\n",
    "    genes[\"gene_id\"] = genes[\"attribute\"].apply(lambda x: extract_attr(x, \"gene_id\"))\n",
    "\n",
    "    genes = genes[[\"seqname\", \"start\", \"end\", \"gene_name\", \"gene_id\"]]\n",
    "    genes.columns = [\"Chromosome\", \"Start\", \"End\", \"gene_name\", \"gene_id\"]\n",
    "\n",
    "    genes[\"Chromosome\"] = genes[\"Chromosome\"].astype(str)\n",
    "    genes[\"Chromosome\"] = genes[\"Chromosome\"].apply(lambda x: x if x.startswith(\"chr\") else f\"chr{x}\")\n",
    "    \n",
    "    return genes\n",
    "\n",
    "genes_df = parse_gtf(\"gencode.v38.annotation.gtf.gz\")\n",
    "gene_ranges = pr.PyRanges(genes_df)\n",
    "\n",
    "CNAs_with_gene = CNAs.join( gene_ranges ).df\n",
    "\n",
    "print(CNAs_with_gene.shape)\n",
    "CNAs_with_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "cbf8b45d-d6cb-462d-aa50-0b7394625be5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "8062 input query terms found dup hits:\t[('WASH7P', 2), ('MIR1302-2HG', 2), ('DDX11L17', 2), ('WASH9P', 2), ('WBP1LP7', 2), ('CICP7', 2), ('\n",
      "20958 input query terms found no hit:\t['RP11-34P13.7', 'RP11-34P13.8', 'RP11-34P13.15', 'RP11-34P13.16', 'RP11-34P13.14', 'RP11-34P13.13',\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(79361, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>653635</td>\n",
       "      <td>8.569510</td>\n",
       "      <td>653635</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>8.569510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>102466751</td>\n",
       "      <td>27.357260</td>\n",
       "      <td>102466751</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>26.562847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>107985730</td>\n",
       "      <td>26.562847</td>\n",
       "      <td>107985730</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79356</th>\n",
       "      <td>ANKRD36P1</td>\n",
       "      <td>ENSG00000188399</td>\n",
       "      <td>8.568661</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79357</th>\n",
       "      <td>ANKRD36P1</td>\n",
       "      <td>100132420</td>\n",
       "      <td>8.568661</td>\n",
       "      <td>100132420</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79358</th>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "      <td>8.401274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79359</th>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000281211</td>\n",
       "      <td>8.401274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79360</th>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>352905</td>\n",
       "      <td>8.401274</td>\n",
       "      <td>352905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79361 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             query              _id     _score entrezgene notfound\n",
       "0           WASH7P           653635   8.569510     653635      NaN\n",
       "1           WASH7P  ENSG00000227232   8.569510        NaN      NaN\n",
       "2        MIR6859-1        102466751  27.357260  102466751      NaN\n",
       "3      MIR1302-2HG  ENSG00000243485  26.562847        NaN      NaN\n",
       "4      MIR1302-2HG        107985730  26.562847  107985730      NaN\n",
       "...            ...              ...        ...        ...      ...\n",
       "79356    ANKRD36P1  ENSG00000188399   8.568661        NaN      NaN\n",
       "79357    ANKRD36P1        100132420   8.568661  100132420      NaN\n",
       "79358      CTBP2P1  ENSG00000235857   8.401274        NaN      NaN\n",
       "79359      CTBP2P1  ENSG00000281211   8.401274        NaN      NaN\n",
       "79360      CTBP2P1           352905   8.401274     352905      NaN\n",
       "\n",
       "[79361 rows x 5 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build Entrez/Gene-name -> HG38 Mapping\n",
    "\n",
    "import shelve\n",
    "import functools\n",
    "import types\n",
    "import mygene\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def wrap_class_method_with_disk_cache(cls, method_name, cache_file):\n",
    "    orig = getattr(cls, method_name)\n",
    "\n",
    "    @functools.wraps(orig)\n",
    "    def wrapper(self, *args, **kwargs):\n",
    "        key = (args, tuple(sorted(kwargs.items())))\n",
    "        key_str = str(key)\n",
    "        with shelve.open(cache_file) as db:\n",
    "            if key_str in db:\n",
    "                return db[key_str]\n",
    "            result = orig(self, *args, **kwargs)\n",
    "            db[key_str] = result\n",
    "            return result\n",
    "\n",
    "    setattr(cls, method_name, wrapper)\n",
    "\n",
    "\n",
    "# import mygene\n",
    "\n",
    "# mg = mygene.MyGeneInfo()\n",
    "# ids = overlap.df['gene_id'].tolist()\n",
    "# res = mg.querymany(ids, scopes='ensembl.gene', fields='entrezgene', species='human')\n",
    "\n",
    "# # entrez_df = pd.DataFrame(res)[[\"query\", \"entrezgene\"]].rename(columns={\"query\": \"gene_id\"})\n",
    "# # final_df = overlap.df.merge(entrez_df, on=\"gene_id\", how=\"left\")\n",
    "\n",
    "\n",
    "wrap_class_method_with_disk_cache(mygene.MyGeneInfo, 'querymany', 'querymany_cache.db')\n",
    "mg = mygene.MyGeneInfo()\n",
    "gene_to_id = mg.querymany(CNAs_with_gene['gene_name'].tolist(), scopes='symbol', fields='entrezgene', species='human')\n",
    "gene_to_id = pd.DataFrame(gene_to_id)\n",
    "\n",
    "print(gene_to_id.shape)\n",
    "gene_to_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "122770a3-6fd3-4afb-acd7-dd5b7a39ebc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(62307, 10)\n",
      "(79361, 5)\n",
      "(62307, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>query</th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.04836</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>653635</td>\n",
       "      <td>8.569510</td>\n",
       "      <td>653635</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.04836</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>102466751</td>\n",
       "      <td>27.357260</td>\n",
       "      <td>102466751</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.04836</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>26.562847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.04836</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>100302278</td>\n",
       "      <td>26.043947</td>\n",
       "      <td>100302278</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.04836</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>645520</td>\n",
       "      <td>27.357826</td>\n",
       "      <td>645520</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GDC_Aliquot_ID Chromosome  Start      End  Num_Probes  \\\n",
       "0  TCGA-A2-A0YM-01A-11D-A894-36       chr1  17001  3723000        2754   \n",
       "1  TCGA-A2-A0YM-01A-11D-A894-36       chr1  17001  3723000        2754   \n",
       "2  TCGA-A2-A0YM-01A-11D-A894-36       chr1  17001  3723000        2754   \n",
       "3  TCGA-A2-A0YM-01A-11D-A894-36       chr1  17001  3723000        2754   \n",
       "4  TCGA-A2-A0YM-01A-11D-A894-36       chr1  17001  3723000        2754   \n",
       "\n",
       "   Segment_Mean  Start_b  End_b    gene_name            gene_id        query  \\\n",
       "0      -0.04836    14404  29570       WASH7P  ENSG00000227232.5       WASH7P   \n",
       "1      -0.04836    17369  17436    MIR6859-1  ENSG00000278267.1    MIR6859-1   \n",
       "2      -0.04836    29554  31109  MIR1302-2HG  ENSG00000243485.5  MIR1302-2HG   \n",
       "3      -0.04836    30366  30503    MIR1302-2  ENSG00000284332.1    MIR1302-2   \n",
       "4      -0.04836    34554  36081      FAM138A  ENSG00000237613.2      FAM138A   \n",
       "\n",
       "               _id     _score entrezgene notfound  \n",
       "0           653635   8.569510     653635      NaN  \n",
       "1        102466751  27.357260  102466751      NaN  \n",
       "2  ENSG00000243485  26.562847        NaN      NaN  \n",
       "3        100302278  26.043947  100302278      NaN  \n",
       "4           645520  27.357826     645520      NaN  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "CNAs_with_gene_dd = CNAs_with_gene.drop_duplicates()\n",
    "gene_to_id_dd = gene_to_id.drop_duplicates(subset=[\"query\"])\n",
    "\n",
    "CNAs_with_entrez = pd.merge(\n",
    "    CNAs_with_gene_dd,\n",
    "    gene_to_id_dd,\n",
    "    left_on='gene_name',\n",
    "    right_on='query',\n",
    "    how='inner'\n",
    ")\n",
    "\n",
    "print(CNAs_with_gene.shape)\n",
    "print(gene_to_id.shape)\n",
    "print(CNAs_with_entrez.shape)\n",
    "df_with_entrez[\"entrezgene\"].isna().sum()\n",
    "\n",
    "# # making sure that this is about to be annoying\n",
    "# (CNAs_with_id.groupby(by=[\"query\"]).apply(lambda x: len(set(x[\"entrezgene\"])) == 1))\n",
    "CNAs_with_entrez.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "9844187f-3fbe-4a63-9b57-48ae4b5488ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "6fe99bf1-9faa-489c-b9f2-fef75d73ee5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['26751' '52' '285016' '129787' '391343' '54221' '7173' '7837' '23040'\n",
      " '730811' '7260' '51112' '55256' '246243' '6201' '78989' '55821' '6664'\n",
      " '150622' '400940' '129607' '91543' '9781' '339788' '3398' '57498'\n",
      " '129642' '8853' '9270' '51692' '285148' '6868' '10971' '9014' '29841'\n",
      " '8462' '192668' '6241' '348738' '3241' '4953' '79954' '245973' '10130'\n",
      " '3754' '130814' '9475' '1876' '9687' '23620' '23175' '28951' '151354'\n",
      " '51594']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "signature_gene_mapping = pd.read_excel(\"copynum_signatures.xlsx\").drop(0).rename(columns={\"Supplementary Data 2. Annotation of copy number segments\": \"Signature\"}).set_index(\"Signature\")\n",
    "\n",
    "# a =  foo[\"Supplementary Data 2. Annotation of copy number segments\"]\n",
    "# output = pd.DataFrame(columns=a)\n",
    "\n",
    "# perfect intersection rarely possible -> introduces error\n",
    "\n",
    "for idx, row in list(signature_gene_mapping.iterrows())[0:1]:\n",
    "    a = row.dropna().apply(lambda x: str(int(x))).to_numpy()\n",
    "\n",
    "    b = CNAs_with_entrez.set_index(\"entrezgene\").index.to_numpy().astype(str)\n",
    "\n",
    "    # c = list(set(a).intersection(set(b)))\n",
    "    \n",
    "    # filtered = CNAs_with_entrez[CNAs_with_entrez[\"entrezgene\"].astype(str).isin(c)]\n",
    "    # print(filtered[\"Segment_Mean\"].hist())\n",
    "\n",
    "    # print(len(a))\n",
    "    # print(len(b))\n",
    "    # print(len(c))\n",
    "    # print(type(a[0]))\n",
    "    # print(type(b[0]))\n",
    "    # print(a)\n",
    "    # print(filtered.head())\n",
    "    # print(type(a))\n",
    "    # print(type(b))\n",
    "    \n",
    "    # row.dropna().to_list()]\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "271dac91-ef2f-401e-b42d-4e2d6e01e6b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>query</th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>notfound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>653635</td>\n",
       "      <td>8.569510</td>\n",
       "      <td>653635</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>102466751</td>\n",
       "      <td>27.357260</td>\n",
       "      <td>102466751</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>26.562847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>100302278</td>\n",
       "      <td>26.043947</td>\n",
       "      <td>100302278</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17001</td>\n",
       "      <td>3723000</td>\n",
       "      <td>2754</td>\n",
       "      <td>-0.048360</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>645520</td>\n",
       "      <td>27.357826</td>\n",
       "      <td>645520</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62302</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>14941746</td>\n",
       "      <td>14942715</td>\n",
       "      <td>AC006989.2</td>\n",
       "      <td>ENSG00000224518.2</td>\n",
       "      <td>AC006989.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62303</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>15348662</td>\n",
       "      <td>15456074</td>\n",
       "      <td>PUDPP1</td>\n",
       "      <td>ENSG00000234620.1</td>\n",
       "      <td>PUDPP1</td>\n",
       "      <td>ENSG00000234620</td>\n",
       "      <td>8.569693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62304</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>11643001</td>\n",
       "      <td>15897000</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.629715</td>\n",
       "      <td>15547216</td>\n",
       "      <td>15593331</td>\n",
       "      <td>STSP1</td>\n",
       "      <td>ENSG00000227166.1</td>\n",
       "      <td>STSP1</td>\n",
       "      <td>6803</td>\n",
       "      <td>8.824922</td>\n",
       "      <td>6803</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62305</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26403001</td>\n",
       "      <td>26442000</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.650024</td>\n",
       "      <td>26409815</td>\n",
       "      <td>26420535</td>\n",
       "      <td>ANKRD36P1</td>\n",
       "      <td>ENSG00000188399.5</td>\n",
       "      <td>ANKRD36P1</td>\n",
       "      <td>ENSG00000188399</td>\n",
       "      <td>8.568661</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62306</th>\n",
       "      <td>TCGA-A2-A0YM-01A-11D-A894-36</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56821001</td>\n",
       "      <td>56882000</td>\n",
       "      <td>61</td>\n",
       "      <td>-0.278835</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000235857.1</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "      <td>8.401274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62307 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     GDC_Aliquot_ID Chromosome     Start       End  \\\n",
       "0      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "1      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "2      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "3      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "4      TCGA-A2-A0YM-01A-11D-A894-36       chr1     17001   3723000   \n",
       "...                             ...        ...       ...       ...   \n",
       "62302  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62303  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62304  TCGA-A2-A0YM-01A-11D-A894-36       chrY  11643001  15897000   \n",
       "62305  TCGA-A2-A0YM-01A-11D-A894-36       chrY  26403001  26442000   \n",
       "62306  TCGA-A2-A0YM-01A-11D-A894-36       chrY  56821001  56882000   \n",
       "\n",
       "       Num_Probes  Segment_Mean   Start_b     End_b    gene_name  \\\n",
       "0            2754     -0.048360     14404     29570       WASH7P   \n",
       "1            2754     -0.048360     17369     17436    MIR6859-1   \n",
       "2            2754     -0.048360     29554     31109  MIR1302-2HG   \n",
       "3            2754     -0.048360     30366     30503    MIR1302-2   \n",
       "4            2754     -0.048360     34554     36081      FAM138A   \n",
       "...           ...           ...       ...       ...          ...   \n",
       "62302          11     -0.629715  14941746  14942715   AC006989.2   \n",
       "62303          11     -0.629715  15348662  15456074       PUDPP1   \n",
       "62304          11     -0.629715  15547216  15593331        STSP1   \n",
       "62305           5     -1.650024  26409815  26420535    ANKRD36P1   \n",
       "62306          61     -0.278835  56855244  56855488      CTBP2P1   \n",
       "\n",
       "                 gene_id        query              _id     _score entrezgene  \\\n",
       "0      ENSG00000227232.5       WASH7P           653635   8.569510     653635   \n",
       "1      ENSG00000278267.1    MIR6859-1        102466751  27.357260  102466751   \n",
       "2      ENSG00000243485.5  MIR1302-2HG  ENSG00000243485  26.562847        NaN   \n",
       "3      ENSG00000284332.1    MIR1302-2        100302278  26.043947  100302278   \n",
       "4      ENSG00000237613.2      FAM138A           645520  27.357826     645520   \n",
       "...                  ...          ...              ...        ...        ...   \n",
       "62302  ENSG00000224518.2   AC006989.2              NaN        NaN        NaN   \n",
       "62303  ENSG00000234620.1       PUDPP1  ENSG00000234620   8.569693        NaN   \n",
       "62304  ENSG00000227166.1        STSP1             6803   8.824922       6803   \n",
       "62305  ENSG00000188399.5    ANKRD36P1  ENSG00000188399   8.568661        NaN   \n",
       "62306  ENSG00000235857.1      CTBP2P1  ENSG00000235857   8.401274        NaN   \n",
       "\n",
       "      notfound  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "...        ...  \n",
       "62302     True  \n",
       "62303      NaN  \n",
       "62304      NaN  \n",
       "62305      NaN  \n",
       "62306      NaN  \n",
       "\n",
       "[62307 rows x 15 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CNAs_with_entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a54c72c-2515-47b3-92da-c8966ec8afcf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
