{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ea8168f4-92f8-4e4d-9b93-41d6e2bcd740",
   "metadata": {},
   "source": [
    "# facts\n",
    "Signatures are continuous segments of chromosomes \\\n",
    "Genes can (and often do) belong to multiple signatures \\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9efecdf-b857-4974-b68b-978bc67074c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "import functools\n",
    "import shelve\n",
    "\n",
    "root = pathlib.Path(\"/data/rbg/shared/datasets/TCGA/\")\n",
    "\n",
    "\n",
    "def wrap_function_with_disk_cache(func, cache_file):\n",
    "    @functools.wraps(func)\n",
    "    def wrapper(*args, **kwargs):\n",
    "        key = (args, tuple(sorted(kwargs.items())))\n",
    "        key_str = str(key)\n",
    "        with shelve.open(cache_file) as db:\n",
    "            if key_str in db:\n",
    "                return db[key_str]\n",
    "            result = func(*args, **kwargs)\n",
    "            db[key_str] = result\n",
    "            return result\n",
    "    return wrapper\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13e383d4-3504-45ec-a358-23d8dc253caf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(133, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project</th>\n",
       "      <th>days_to_consent</th>\n",
       "      <th>diagnoses</th>\n",
       "      <th>consent_type</th>\n",
       "      <th>demographic</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>case_id</th>\n",
       "      <th>follow_ups</th>\n",
       "      <th>index_date</th>\n",
       "      <th>state</th>\n",
       "      <th>lost_to_followup</th>\n",
       "      <th>exposures</th>\n",
       "      <th>outcome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[{'tissue_or_organ_of_origin': 'Breast, NOS', ...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:56.031656-06:00</td>\n",
       "      <td>a9bb8159-32f0-454c-a946-b3286a52b9d5</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 584]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>22.0</td>\n",
       "      <td>[{'morphology': '8500/3', 'submitter_id': 'TCG...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '92faa22e-8e19-5502-987c-a6...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:38:16.470569-06:00</td>\n",
       "      <td>ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21</td>\n",
       "      <td>[{'days_to_progression': 102, 'timepoint_categ...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 548]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A04T</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T21:21:42.777520-06:00</td>\n",
       "      <td>b58ad350-5140-4fa8-bc2c-24bca8395f3a</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2246]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '9145b76c-61b5-56e6-876e-58...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-05T23:51:14.306282-06:00</td>\n",
       "      <td>1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9</td>\n",
       "      <td>[{'timepoint_category': 'Follow-up', 'follow_u...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 2654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>{'project_id': 'TCGA-BRCA'}</td>\n",
       "      <td>-212.0</td>\n",
       "      <td>[{'synchronous_malignancy': 'No', 'ajcc_pathol...</td>\n",
       "      <td>Informed Consent</td>\n",
       "      <td>{'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...</td>\n",
       "      <td>Breast</td>\n",
       "      <td>2025-01-06T07:33:56.519794-06:00</td>\n",
       "      <td>eb2dbb4f-66b6-4525-8323-431970f7a64e</td>\n",
       "      <td>[{'timepoint_category': 'Last Contact', 'follo...</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>released</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1, 754]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   submitter_id                  disease_type                      project  \\\n",
       "0  TCGA-A1-A0SP  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "1  TCGA-A2-A04P  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "2  TCGA-A2-A04T  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "3  TCGA-A2-A04U  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "4  TCGA-A2-A0CM  Ductal and Lobular Neoplasms  {'project_id': 'TCGA-BRCA'}   \n",
       "\n",
       "   days_to_consent                                          diagnoses  \\\n",
       "0             55.0  [{'tissue_or_organ_of_origin': 'Breast, NOS', ...   \n",
       "1             22.0  [{'morphology': '8500/3', 'submitter_id': 'TCG...   \n",
       "2             29.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "3             23.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "4           -212.0  [{'synchronous_malignancy': 'No', 'ajcc_pathol...   \n",
       "\n",
       "       consent_type                                        demographic  \\\n",
       "0  Informed Consent  {'demographic_id': 'f7ed8691-ed3d-54bf-94c5-7f...   \n",
       "1  Informed Consent  {'demographic_id': '92faa22e-8e19-5502-987c-a6...   \n",
       "2  Informed Consent  {'demographic_id': 'dd68f2a1-f279-567f-83b6-4c...   \n",
       "3  Informed Consent  {'demographic_id': '9145b76c-61b5-56e6-876e-58...   \n",
       "4  Informed Consent  {'demographic_id': '0f8728a4-3abb-5a78-84c3-3d...   \n",
       "\n",
       "  primary_site                  updated_datetime  \\\n",
       "0       Breast  2025-01-06T07:38:56.031656-06:00   \n",
       "1       Breast  2025-01-06T07:38:16.470569-06:00   \n",
       "2       Breast  2025-01-05T21:21:42.777520-06:00   \n",
       "3       Breast  2025-01-05T23:51:14.306282-06:00   \n",
       "4       Breast  2025-01-06T07:33:56.519794-06:00   \n",
       "\n",
       "                                case_id  \\\n",
       "0  a9bb8159-32f0-454c-a946-b3286a52b9d5   \n",
       "1  ccd4a24b-d8cc-4686-9dee-c98b0c5a8d21   \n",
       "2  b58ad350-5140-4fa8-bc2c-24bca8395f3a   \n",
       "3  1c3610f7-e0aa-48d7-9a27-0dbaf6e244f9   \n",
       "4  eb2dbb4f-66b6-4525-8323-431970f7a64e   \n",
       "\n",
       "                                          follow_ups index_date     state  \\\n",
       "0  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "1  [{'days_to_progression': 102, 'timepoint_categ...  Diagnosis  released   \n",
       "2  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "3  [{'timepoint_category': 'Follow-up', 'follow_u...  Diagnosis  released   \n",
       "4  [{'timepoint_category': 'Last Contact', 'follo...  Diagnosis  released   \n",
       "\n",
       "  lost_to_followup exposures    outcome  \n",
       "0              NaN       NaN   [0, 584]  \n",
       "1               No       NaN   [1, 548]  \n",
       "2              NaN       NaN  [0, 2246]  \n",
       "3              Yes       NaN  [0, 2654]  \n",
       "4              NaN       NaN   [1, 754]  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def add_outcome(elem):\n",
    "    vital_status = elem['demographic'].get('vital_status')\n",
    "\n",
    "    if vital_status == 'Dead':\n",
    "        patient_death = elem['demographic'].get('days_to_death')\n",
    "        return [1, int(patient_death)]\n",
    "\n",
    "    elif vital_status == \"Alive\":\n",
    "        follow_ups = elem['follow_ups']\n",
    "        follow_up_days = [\n",
    "            f.get('days_to_follow_up')\n",
    "            for f in follow_ups\n",
    "            if f.get('days_to_follow_up') is not None\n",
    "        ]\n",
    "        diagnosis_days = [\n",
    "            d.get('days_to_last_follow_up')\n",
    "            for d in elem['diagnoses']\n",
    "            if d.get('days_to_last_follow_up') is not None\n",
    "        ]\n",
    "        all_days = follow_up_days + diagnosis_days\n",
    "        date_of_last_checkup = max(all_days) if all_days else None\n",
    "\n",
    "        return [0, int(date_of_last_checkup)]\n",
    "    else:\n",
    "        typing.assert_never(\"vital status not in Alive/Dead\")\n",
    "\n",
    "\n",
    "def remove_sparse_features(X, threshold=0.05):\n",
    "    non_zero_fraction = (X != 0).sum(axis=0) / X.shape[0]\n",
    "    keep = non_zero_fraction > threshold\n",
    "    return X.loc[:, keep]\n",
    "\n",
    "\n",
    "def remove_perfect_correlations(X):\n",
    "    corr_matrix = X.corr().abs()\n",
    "    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "    to_drop = [column for column in upper.columns if any(upper[column] == 1)]\n",
    "    return X.drop(columns=to_drop)\n",
    "\n",
    "\n",
    "# ids = pd.read_json(\"ids/tcga_cohort.json\")[\"case_id\"]\n",
    "ids = pd.read_csv(\"sample_ids.csv\")[\"Sample\"]\n",
    "\n",
    "case_ids_raw = pd.read_json(\"./clinical-tcga-brca.json\").set_index(\"submitter_id\")\n",
    "case_ids = case_ids_raw.loc[ids]\n",
    "case_ids[\"outcome\"] = case_ids.apply(add_outcome, axis=1)\n",
    "\n",
    "case_ids = case_ids.reset_index()\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids = case_ids.drop_duplicates(subset=[\"case_id\"])\n",
    "print(case_ids.shape)\n",
    "\n",
    "case_ids.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f255396-1170-4260-ab44-34755d0bae5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2774, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Tumor Descriptor</th>\n",
       "      <th>Specimen Type</th>\n",
       "      <th>Preservation Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0518551d-4df2-4124-b68d-494200c5586b</td>\n",
       "      <td>e75893cc-346b-4567-bdd4-f88f3444a72a.wxs.aliqu...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Somatic Mutation</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-E2-A158</td>\n",
       "      <td>TCGA-E2-A158-01A, TCGA-E2-A158-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>Unknown, Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98953412-ae65-4116-b43e-20278294e4f2</td>\n",
       "      <td>TCGA_BRCA.a5279f5e-626b-45ca-86fa-976ae328fa2b...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Annotated Somatic Mutation</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-E9-A1R0</td>\n",
       "      <td>TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3acd8060-5f26-4d22-b318-1b2427e1e30c</td>\n",
       "      <td>54f9af00-9550-45a5-b3c4-3af171e0300a.wxs.VarSc...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Annotated Somatic Mutation</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-C8-A27B</td>\n",
       "      <td>TCGA-C8-A27B-10A, TCGA-C8-A27B-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Solid Tissue</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4a3fc24a-c86b-48e0-bc11-c91fcc09a317</td>\n",
       "      <td>308c1b03-8cf4-44a5-aa18-12f9e965d85b.wxs.aliqu...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Somatic Mutation</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-AN-A03Y</td>\n",
       "      <td>TCGA-AN-A03Y-10A, TCGA-AN-A03Y-01A</td>\n",
       "      <td>Normal, Tumor</td>\n",
       "      <td>Not Applicable, Primary</td>\n",
       "      <td>Peripheral Blood NOS, Unknown</td>\n",
       "      <td>Unknown, OCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>938979df-2611-4826-9982-8cfc0395e5dd</td>\n",
       "      <td>efb68702-7e0e-474d-a6b5-2855d963e3da.wxs.aliqu...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Aggregated Somatic Mutation</td>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-B6-A0WW</td>\n",
       "      <td>TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A</td>\n",
       "      <td>Tumor, Normal</td>\n",
       "      <td>Primary, Not Applicable</td>\n",
       "      <td>Solid Tissue, Peripheral Blood NOS</td>\n",
       "      <td>OCT, Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  0518551d-4df2-4124-b68d-494200c5586b   \n",
       "2  98953412-ae65-4116-b43e-20278294e4f2   \n",
       "3  3acd8060-5f26-4d22-b318-1b2427e1e30c   \n",
       "5  4a3fc24a-c86b-48e0-bc11-c91fcc09a317   \n",
       "6  938979df-2611-4826-9982-8cfc0395e5dd   \n",
       "\n",
       "                                           File Name  \\\n",
       "0  e75893cc-346b-4567-bdd4-f88f3444a72a.wxs.aliqu...   \n",
       "2  TCGA_BRCA.a5279f5e-626b-45ca-86fa-976ae328fa2b...   \n",
       "3  54f9af00-9550-45a5-b3c4-3af171e0300a.wxs.VarSc...   \n",
       "5  308c1b03-8cf4-44a5-aa18-12f9e965d85b.wxs.aliqu...   \n",
       "6  efb68702-7e0e-474d-a6b5-2855d963e3da.wxs.aliqu...   \n",
       "\n",
       "                 Data Category                    Data Type Project ID  \\\n",
       "0  Simple Nucleotide Variation      Masked Somatic Mutation  TCGA-BRCA   \n",
       "2  Simple Nucleotide Variation   Annotated Somatic Mutation  TCGA-BRCA   \n",
       "3  Simple Nucleotide Variation   Annotated Somatic Mutation  TCGA-BRCA   \n",
       "5  Simple Nucleotide Variation      Masked Somatic Mutation  TCGA-BRCA   \n",
       "6  Simple Nucleotide Variation  Aggregated Somatic Mutation  TCGA-BRCA   \n",
       "\n",
       "        Case ID                           Sample ID    Tissue Type  \\\n",
       "0  TCGA-E2-A158  TCGA-E2-A158-01A, TCGA-E2-A158-10A  Tumor, Normal   \n",
       "2  TCGA-E9-A1R0  TCGA-E9-A1R0-10A, TCGA-E9-A1R0-01A  Normal, Tumor   \n",
       "3  TCGA-C8-A27B  TCGA-C8-A27B-10A, TCGA-C8-A27B-01A  Normal, Tumor   \n",
       "5  TCGA-AN-A03Y  TCGA-AN-A03Y-10A, TCGA-AN-A03Y-01A  Normal, Tumor   \n",
       "6  TCGA-B6-A0WW  TCGA-B6-A0WW-01A, TCGA-B6-A0WW-10A  Tumor, Normal   \n",
       "\n",
       "          Tumor Descriptor                       Specimen Type  \\\n",
       "0  Primary, Not Applicable  Solid Tissue, Peripheral Blood NOS   \n",
       "2  Not Applicable, Primary  Peripheral Blood NOS, Solid Tissue   \n",
       "3  Not Applicable, Primary  Peripheral Blood NOS, Solid Tissue   \n",
       "5  Not Applicable, Primary       Peripheral Blood NOS, Unknown   \n",
       "6  Primary, Not Applicable  Solid Tissue, Peripheral Blood NOS   \n",
       "\n",
       "  Preservation Method  \n",
       "0    Unknown, Unknown  \n",
       "2        Unknown, OCT  \n",
       "3        Unknown, OCT  \n",
       "5        Unknown, OCT  \n",
       "6        OCT, Unknown  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_mapping = \"/data/rbg/shared/datasets/TCGA/Genomic/mutations.gdc_sample_sheet.2025-07-07.tsv\"\n",
    "\n",
    "caseid_mut_mapping = pd.read_csv(dna_mapping, delimiter='\\t')\n",
    "\n",
    "caseid_mut_mapping['Case ID'] = caseid_mut_mapping['Case ID'].str.split(',')\n",
    "# print(caseid_mut_mapping.shape)\n",
    "caseid_mut_mapping = caseid_mut_mapping.explode('Case ID')\n",
    "caseid_mut_mapping['Case ID'] = caseid_mut_mapping['Case ID'].str.strip()\n",
    "# print(caseid_mut_mapping.shape)\n",
    "caseid_mut_mapping = caseid_mut_mapping.drop_duplicates(subset=[\"Case ID\"]) # I'm pretty sure this is valid bc I explode only on Case ID - meaning everything else is duplicated. This whole process might actualyl be unncesary. TODO: check if all \"Case IDs pairs\" are just twins\n",
    "print(caseid_mut_mapping.shape)\n",
    "\n",
    "\n",
    "# # stupid test to check whether any of hte records had duplicate sample IDs\n",
    "# # def foo(x):\n",
    "# #     a = list(map(lambda x: x.strip(), x[\"Sample ID\"].split(\",\")))\n",
    "# #     print(a[0])\n",
    "# #     print(a[1])\n",
    "# #     # return a[0] == a[1]\n",
    "# # caseid_copynum_mapping = caseid_copynum_mapping[caseid_copynum_mapping.apply(foo, axis=1)]\n",
    "\n",
    "caseid_mut_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c5ae89c6-a920-4472-af53-a156e7b4e45f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(133, 16)\n",
      "(133, 20)\n"
     ]
    }
   ],
   "source": [
    "# add copynumbesr to \"IDs\" dataset\n",
    "# FIXME: figure out why 21 records are being dropped - the final result should still be 133\n",
    "print(case_ids.shape)\n",
    "data= pd.merge(\n",
    "    case_ids,\n",
    "    caseid_mut_mapping[['Case ID', 'Sample ID', 'File ID', \"File Name\"]],  # select only the needed column\n",
    "    left_on='submitter_id',\n",
    "    right_on='Case ID',\n",
    "    how=\"inner\"\n",
    ")\n",
    "print(data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bee10e64-a3e0-4a3f-961e-ad6ae84dd330",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['logs',\n",
       " 'c30ef13b-f23a-46c2-9187-f64269318358.wxs.pindel.raw_somatic_mutation.vcf.gz',\n",
       " 'c30ef13b-f23a-46c2-9187-f64269318358.wxs.pindel.raw_somatic_mutation.vcf.gz.tbi']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# !pip install BGZF\n",
    "\n",
    "# !zcat /data/rbg/shared/datasets/TCGA/Genomic/downloads/49fe1ce1-3dc8-4130-a29c-3f3ca82bfe61/c30ef13b-f23a-46c2-9187-f64269318358.wxs.pindel.raw_somatic_mutation.vcf.gz | head -20\n",
    "# !file /data/rbg/shared/datasets/TCGA/Genomic/downloads/49fe1ce1-3dc8-4130-a29c-3f3ca82bfe61/c30ef13b-f23a-46c2-9187-f64269318358.wxs.pindel.raw_somatic_mutation.vcf.gz\n",
    "\n",
    "os.listdir(f\"{root}/Genomic/downloads/{elem['File ID']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "fd3ff3c0-13cf-45ed-8a49-3bf869824169",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/rbg/shared/datasets/TCGA/Genomic/downloads/dd7e8bb1-e084-4b8a-bf3b-1064de7444d8/TCGA_BRCA.3c9305f7-9ef0-44f4-8caa-c73e689ad319.wxs.MuSE.somatic_annotation.vcf.gz\n",
      "/data/rbg/shared/datasets/TCGA/Genomic/downloads/dd7e8bb1-e084-4b8a-bf3b-1064de7444d8/TCGA_BRCA.3c9305f7-9ef0-44f4-8caa-c73e689ad319.wxs.MuSE.somatic_annotation.vcf.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filters</th>\n",
       "      <th>somatic</th>\n",
       "      <th>consequence</th>\n",
       "      <th>impact</th>\n",
       "      <th>gene</th>\n",
       "      <th>clin_sig</th>\n",
       "      <th>normal_gt</th>\n",
       "      <th>tumor_gt</th>\n",
       "      <th>normal_dp</th>\n",
       "      <th>tumor_dp</th>\n",
       "      <th>normal_ad</th>\n",
       "      <th>tumor_ad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1022338</td>\n",
       "      <td>None</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>[PASS]</td>\n",
       "      <td>True</td>\n",
       "      <td>A|5_prime_UTR_variant|MODIFIER|AGRN|ENSG000001...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(0, 0)</td>\n",
       "      <td>(0, 1)</td>\n",
       "      <td>19</td>\n",
       "      <td>170</td>\n",
       "      <td>(19, 0)</td>\n",
       "      <td>(119, 50)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1750064</td>\n",
       "      <td>None</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>[PASS]</td>\n",
       "      <td>True</td>\n",
       "      <td>A|downstream_gene_variant|MODIFIER|NADK|ENSG00...</td>\n",
       "      <td>A|downstream_gene_variant|MODIFIER|NADK|ENSG00...</td>\n",
       "      <td>A|downstream_gene_variant|MODIFIER|NADK|ENSG00...</td>\n",
       "      <td>None</td>\n",
       "      <td>(0, 0)</td>\n",
       "      <td>(0, 1)</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>(8, 0)</td>\n",
       "      <td>(9, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2390118</td>\n",
       "      <td>None</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>[PASS]</td>\n",
       "      <td>True</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|RER1|ENSG0000...</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|RER1|ENSG0000...</td>\n",
       "      <td>T|intron_variant&amp;NMD_transcript_variant|MODIFI...</td>\n",
       "      <td>None</td>\n",
       "      <td>(0, 0)</td>\n",
       "      <td>(0, 1)</td>\n",
       "      <td>34</td>\n",
       "      <td>51</td>\n",
       "      <td>(34, 0)</td>\n",
       "      <td>(35, 16)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2652391</td>\n",
       "      <td>None</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>None</td>\n",
       "      <td>[Tier1]</td>\n",
       "      <td>True</td>\n",
       "      <td>C|intron_variant|MODIFIER|TTC34|ENSG0000021591...</td>\n",
       "      <td>C|regulatory_region_variant|MODIFIER|||Regulat...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(0, 0)</td>\n",
       "      <td>(0, 1)</td>\n",
       "      <td>30</td>\n",
       "      <td>47</td>\n",
       "      <td>(29, 1)</td>\n",
       "      <td>(45, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2653973</td>\n",
       "      <td>None</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>[PASS]</td>\n",
       "      <td>True</td>\n",
       "      <td>G|intron_variant|MODIFIER|TTC34|ENSG0000021591...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(0, 0)</td>\n",
       "      <td>(0, 1)</td>\n",
       "      <td>70</td>\n",
       "      <td>132</td>\n",
       "      <td>(69, 0)</td>\n",
       "      <td>(124, 8)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      pos    id ref alt  qual  filters  somatic  \\\n",
       "0  chr1  1022338  None   G   A  None   [PASS]     True   \n",
       "1  chr1  1750064  None   T   A  None   [PASS]     True   \n",
       "2  chr1  2390118  None   C   T  None   [PASS]     True   \n",
       "3  chr1  2652391  None   G   C  None  [Tier1]     True   \n",
       "4  chr1  2653973  None   C   G  None   [PASS]     True   \n",
       "\n",
       "                                         consequence  \\\n",
       "0  A|5_prime_UTR_variant|MODIFIER|AGRN|ENSG000001...   \n",
       "1  A|downstream_gene_variant|MODIFIER|NADK|ENSG00...   \n",
       "2  T|upstream_gene_variant|MODIFIER|RER1|ENSG0000...   \n",
       "3  C|intron_variant|MODIFIER|TTC34|ENSG0000021591...   \n",
       "4  G|intron_variant|MODIFIER|TTC34|ENSG0000021591...   \n",
       "\n",
       "                                              impact  \\\n",
       "0                                               None   \n",
       "1  A|downstream_gene_variant|MODIFIER|NADK|ENSG00...   \n",
       "2  T|upstream_gene_variant|MODIFIER|RER1|ENSG0000...   \n",
       "3  C|regulatory_region_variant|MODIFIER|||Regulat...   \n",
       "4                                               None   \n",
       "\n",
       "                                                gene clin_sig normal_gt  \\\n",
       "0                                               None     None    (0, 0)   \n",
       "1  A|downstream_gene_variant|MODIFIER|NADK|ENSG00...     None    (0, 0)   \n",
       "2  T|intron_variant&NMD_transcript_variant|MODIFI...     None    (0, 0)   \n",
       "3                                               None     None    (0, 0)   \n",
       "4                                               None     None    (0, 0)   \n",
       "\n",
       "  tumor_gt  normal_dp  tumor_dp normal_ad   tumor_ad  \n",
       "0   (0, 1)         19       170   (19, 0)  (119, 50)  \n",
       "1   (0, 1)          8        12    (8, 0)     (9, 3)  \n",
       "2   (0, 1)         34        51   (34, 0)   (35, 16)  \n",
       "3   (0, 1)         30        47   (29, 1)    (45, 2)  \n",
       "4   (0, 1)         70       132   (69, 0)   (124, 8)  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pysam\n",
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import requests\n",
    "import time\n",
    "\n",
    "elem = data.iloc[8]\n",
    "path = f\"{root}/Genomic/downloads/{elem['File ID']}/{elem['File Name']}\"\n",
    "print(path)\n",
    "assert (path.endswith(\".gz\"))\n",
    "\n",
    "vcf = pysam.VariantFile(path)\n",
    "\n",
    "import pysam\n",
    "import pandas as pd\n",
    "\n",
    "elem = data.iloc[8]\n",
    "path = f\"{root}/Genomic/downloads/{elem['File ID']}/{elem['File Name']}\"\n",
    "print(path)\n",
    "assert path.endswith(\".gz\")\n",
    "\n",
    "vcf = pysam.VariantFile(path)\n",
    "\n",
    "variants = []\n",
    "\n",
    "for record in vcf.fetch():\n",
    "    chrom = record.chrom\n",
    "    pos = record.pos\n",
    "    var_id = record.id\n",
    "    ref = record.ref\n",
    "    alt = ','.join(str(a) for a in record.alts)\n",
    "    qual = record.qual\n",
    "    \n",
    "    filters = list(record.filter.keys()) if record.filter else []\n",
    "    \n",
    "    # INFO fields - somatic flag, CSQ annotation, allele counts\n",
    "    is_somatic = 'SOMATIC' in record.info\n",
    "    \n",
    "    csq = record.info.get('CSQ')\n",
    "    # csq_fields = csq.split(',')[0].split('|') if csq else []\n",
    "    csq_fields = list(csq)\n",
    "    \n",
    "    # Extract some common VEP fields by their position in the CSQ format:\n",
    "    # Format is documented in the header, but common fields: \n",
    "    # Allele(0),Consequence(1),IMPACT(2),SYMBOL(3),Gene(4),Feature_type(5),Feature(6),BIOTYPE(7),HGVSc(10),HGVSp(11),CLIN_SIG(66) etc.\n",
    "    # Adjust indices if needed based on your actual header\n",
    "    consequence = csq_fields[1] if len(csq_fields) > 1 else None\n",
    "    impact = csq_fields[2] if len(csq_fields) > 2 else None\n",
    "    gene = csq_fields[4] if len(csq_fields) > 4 else None\n",
    "    clin_sig = csq_fields[66] if len(csq_fields) > 66 else None\n",
    "    \n",
    "    # FORMAT fields for each sample\n",
    "    # Assume samples named \"NORMAL\" and \"TUMOR\" as per header\n",
    "    normal_sample = record.samples.get('NORMAL')\n",
    "    tumor_sample = record.samples.get('TUMOR')\n",
    "    \n",
    "    # Genotype\n",
    "    normal_gt = normal_sample['GT'] if normal_sample else None\n",
    "    tumor_gt = tumor_sample['GT'] if tumor_sample else None\n",
    "    \n",
    "    normal_dp = normal_sample['DP'] if normal_sample and 'DP' in normal_sample else None\n",
    "    tumor_dp = tumor_sample['DP'] if tumor_sample and 'DP' in tumor_sample else None\n",
    "    \n",
    "    # Allelic depths (AD), tuple or list per REF and ALT\n",
    "    normal_ad = normal_sample['AD'] if normal_sample and 'AD' in normal_sample else None\n",
    "    tumor_ad = tumor_sample['AD'] if tumor_sample and 'AD' in tumor_sample else None\n",
    "    \n",
    "    variants.append({\n",
    "        'chrom': chrom,\n",
    "        'pos': pos,\n",
    "        'id': var_id,\n",
    "        'ref': ref,\n",
    "        'alt': alt,\n",
    "        'qual': qual,\n",
    "        'filters': filters,\n",
    "        'somatic': is_somatic,\n",
    "        'consequence': consequence,\n",
    "        'impact': impact,\n",
    "        'gene': gene,\n",
    "        'clin_sig': clin_sig,\n",
    "        'normal_gt': normal_gt,\n",
    "        'tumor_gt': tumor_gt,\n",
    "        'normal_dp': normal_dp,\n",
    "        'tumor_dp': tumor_dp,\n",
    "        'normal_ad': normal_ad,\n",
    "        'tumor_ad': tumor_ad,\n",
    "    })\n",
    "\n",
    "variants_df = pd.DataFrame(variants)\n",
    "variants_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "7e15a39c-cb91-4621-95e8-c88019343f8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/rbg/shared/datasets/TCGA/Genomic/downloads/dd7e8bb1-e084-4b8a-bf3b-1064de7444d8/TCGA_BRCA.3c9305f7-9ef0-44f4-8caa-c73e689ad319.wxs.MuSE.somatic_annotation.vcf.gz: Blocked GNU Zip Format (BGZF; gzip compatible), block length 11659\n",
      "['logs', 'TCGA_BRCA.3c9305f7-9ef0-44f4-8caa-c73e689ad319.wxs.MuSE.somatic_annotation.vcf.gz', 'TCGA_BRCA.3c9305f7-9ef0-44f4-8caa-c73e689ad319.wxs.MuSE.somatic_annotation.vcf.gz.tbi']\n"
     ]
    }
   ],
   "source": [
    "!file $path\n",
    "print(os.listdir(f\"{root}/Genomic/downloads/{elem['File ID']}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f3046759-c27c-44e1-87b1-aaa37dc84292",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Pathlib' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[76]\u001b[39m\u001b[32m, line 47\u001b[39m\n\u001b[32m     35\u001b[39m         \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mvcf2maf failed: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m     37\u001b[39m \u001b[38;5;66;03m# export VCF2MAF_URL=`curl -sL https://api.github.com/repos/mskcc/vcf2maf/releases | grep -m1 tarball_url | cut -d\\\" -f4`\u001b[39;00m\n\u001b[32m     38\u001b[39m \u001b[38;5;66;03m# curl -L -o mskcc-vcf2maf.tar.gz $VCF2MAF_URL; tar -zxf mskcc-vcf2maf.tar.gz; cd mskcc-vcf2maf-*\u001b[39;00m\n\u001b[32m     39\u001b[39m \u001b[38;5;66;03m# mkdir vcf2maf_cache\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     43\u001b[39m \u001b[38;5;66;03m# cpan DBI\u001b[39;00m\n\u001b[32m     44\u001b[39m \u001b[38;5;66;03m# perl INSTALL.pl\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m47\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[43mPathlib\u001b[49m.Path(path).with_suffix(\u001b[33m\"\u001b[39m\u001b[33mmaf\u001b[39m\u001b[33m\"\u001b[39m))\n\u001b[32m     49\u001b[39m \u001b[38;5;66;03m# if __name__ == \"__main__\":\u001b[39;00m\n\u001b[32m     50\u001b[39m \u001b[38;5;66;03m#     vcf_path = path\u001b[39;00m\n\u001b[32m     51\u001b[39m \u001b[38;5;66;03m#     maf_output = \u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     55\u001b[39m \n\u001b[32m     56\u001b[39m \u001b[38;5;66;03m#     run_vcf2maf(vcf_path, maf_output, ref_fasta, vep_path, vep_cache)\u001b[39;00m\n",
      "\u001b[31mNameError\u001b[39m: name 'Pathlib' is not defined"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import pathlib\n",
    "import os\n",
    "\n",
    "def run_vcf2maf(vcf_path, maf_output, ref_fasta, vep_path, vep_cache):\n",
    "    \"\"\"\n",
    "    Convert VCF to MAF using vcf2maf.pl wrapper.\n",
    "\n",
    "    Args:\n",
    "        vcf_path (str): Input VCF file path (gzipped recommended).\n",
    "        maf_output (str): Output MAF file path.\n",
    "        ref_fasta (str): Path to reference fasta (e.g. GRCh38.fa).\n",
    "        vep_path (str): Path to VEP executable.\n",
    "        vep_cache (str): Path to VEP cache directory.\n",
    "    \"\"\"\n",
    "    # Command template for vcf2maf.pl\n",
    "    cmd = [\n",
    "        \"vcf2maf.pl\",\n",
    "        \"--input-vcf\", vcf_path,\n",
    "        \"--output-maf\", maf_output,\n",
    "        \"--ref-fasta\", ref_fasta,\n",
    "        \"--vep-path\", vep_path,\n",
    "        \"--vep-data\", vep_cache,\n",
    "        \"--tumor-id\", \"TUMOR\",      # Adjust sample IDs as needed\n",
    "        \"--normal-id\", \"NORMAL\"\n",
    "    ]\n",
    "\n",
    "    print(\"Running vcf2maf with command:\")\n",
    "    print(\" \".join(cmd))\n",
    "\n",
    "    try:\n",
    "        subprocess.run(cmd, check=True)\n",
    "        print(f\"MAF successfully generated: {maf_output}\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"vcf2maf failed: {e}\")\n",
    "\n",
    "# export VCF2MAF_URL=`curl -sL https://api.github.com/repos/mskcc/vcf2maf/releases | grep -m1 tarball_url | cut -d\\\" -f4`\n",
    "# curl -L -o mskcc-vcf2maf.tar.gz $VCF2MAF_URL; tar -zxf mskcc-vcf2maf.tar.gz; cd mskcc-vcf2maf-*\n",
    "# mkdir vcf2maf_cache\n",
    "\n",
    "# git clone https://github.com/Ensembl/ensembl-vep.git\n",
    "# cd ensemble-vep.git\n",
    "# cpan DBI\n",
    "# perl INSTALL.pl\n",
    "\n",
    "\n",
    "print(pathlib.Path(path).with_suffix(\"maf\"))\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     vcf_path = path\n",
    "#     maf_output = \n",
    "#     ref_fasta = \"GRCh38.primary_assembly.genome.fa.gz\"\n",
    "#     vep_path = \"./vcf2maf/vcf2maf.pl\"\n",
    "#     vep_cache = \"./vcf2maf_cache/\"\n",
    "\n",
    "#     run_vcf2maf(vcf_path, maf_output, ref_fasta, vep_path, vep_cache)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2be5e3-fe23-4b5c-9796-299f7210f8ae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
